{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db21bbdb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pypdf in a:\\work\\langchain\\lang\\lib\\site-packages (6.0.0)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[notice] A new release of pip is available: 25.1.1 -> 25.2\n",
      "[notice] To update, run: python.exe -m pip install --upgrade pip\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pdfplumber in a:\\work\\langchain\\lang\\lib\\site-packages (0.11.7)\n",
      "Requirement already satisfied: pdfminer.six==20250506 in a:\\work\\langchain\\lang\\lib\\site-packages (from pdfplumber) (20250506)\n",
      "Requirement already satisfied: Pillow>=9.1 in a:\\work\\langchain\\lang\\lib\\site-packages (from pdfplumber) (11.2.1)\n",
      "Requirement already satisfied: pypdfium2>=4.18.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from pdfplumber) (4.30.0)\n",
      "Requirement already satisfied: charset-normalizer>=2.0.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from pdfminer.six==20250506->pdfplumber) (3.4.2)\n",
      "Requirement already satisfied: cryptography>=36.0.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from pdfminer.six==20250506->pdfplumber) (45.0.6)\n",
      "Requirement already satisfied: cffi>=1.14 in a:\\work\\langchain\\lang\\lib\\site-packages (from cryptography>=36.0.0->pdfminer.six==20250506->pdfplumber) (1.17.1)\n",
      "Requirement already satisfied: pycparser in a:\\work\\langchain\\lang\\lib\\site-packages (from cffi>=1.14->cryptography>=36.0.0->pdfminer.six==20250506->pdfplumber) (2.22)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[notice] A new release of pip is available: 25.1.1 -> 25.2\n",
      "[notice] To update, run: python.exe -m pip install --upgrade pip\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting chromadb\n",
      "  Downloading chromadb-1.0.17-cp39-abi3-win_amd64.whl.metadata (7.4 kB)\n",
      "Collecting build>=1.0.3 (from chromadb)\n",
      "  Downloading build-1.3.0-py3-none-any.whl.metadata (5.6 kB)\n",
      "Requirement already satisfied: pydantic>=1.9 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (2.11.4)\n",
      "Collecting pybase64>=1.4.1 (from chromadb)\n",
      "  Downloading pybase64-1.4.2-cp313-cp313-win_amd64.whl.metadata (9.0 kB)\n",
      "Collecting uvicorn>=0.18.3 (from uvicorn[standard]>=0.18.3->chromadb)\n",
      "  Downloading uvicorn-0.35.0-py3-none-any.whl.metadata (6.5 kB)\n",
      "Requirement already satisfied: numpy>=1.22.5 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (2.2.5)\n",
      "Collecting posthog<6.0.0,>=2.4.0 (from chromadb)\n",
      "  Downloading posthog-5.4.0-py3-none-any.whl.metadata (5.7 kB)\n",
      "Requirement already satisfied: typing-extensions>=4.5.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (4.13.2)\n",
      "Collecting onnxruntime>=1.14.1 (from chromadb)\n",
      "  Downloading onnxruntime-1.22.1-cp313-cp313-win_amd64.whl.metadata (5.1 kB)\n",
      "Collecting opentelemetry-api>=1.2.0 (from chromadb)\n",
      "  Downloading opentelemetry_api-1.36.0-py3-none-any.whl.metadata (1.5 kB)\n",
      "Collecting opentelemetry-exporter-otlp-proto-grpc>=1.2.0 (from chromadb)\n",
      "  Downloading opentelemetry_exporter_otlp_proto_grpc-1.36.0-py3-none-any.whl.metadata (2.4 kB)\n",
      "Collecting opentelemetry-sdk>=1.2.0 (from chromadb)\n",
      "  Downloading opentelemetry_sdk-1.36.0-py3-none-any.whl.metadata (1.5 kB)\n",
      "Requirement already satisfied: tokenizers>=0.13.2 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (0.21.1)\n",
      "Collecting pypika>=0.48.9 (from chromadb)\n",
      "  Downloading PyPika-0.48.9.tar.gz (67 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: tqdm>=4.65.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (4.67.1)\n",
      "Collecting overrides>=7.3.1 (from chromadb)\n",
      "  Using cached overrides-7.7.0-py3-none-any.whl.metadata (5.8 kB)\n",
      "Collecting importlib-resources (from chromadb)\n",
      "  Downloading importlib_resources-6.5.2-py3-none-any.whl.metadata (3.9 kB)\n",
      "Requirement already satisfied: grpcio>=1.58.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (1.71.0)\n",
      "Collecting bcrypt>=4.0.1 (from chromadb)\n",
      "  Downloading bcrypt-4.3.0-cp39-abi3-win_amd64.whl.metadata (10 kB)\n",
      "Collecting typer>=0.9.0 (from chromadb)\n",
      "  Downloading typer-0.16.0-py3-none-any.whl.metadata (15 kB)\n",
      "Collecting kubernetes>=28.1.0 (from chromadb)\n",
      "  Downloading kubernetes-33.1.0-py2.py3-none-any.whl.metadata (1.7 kB)\n",
      "Requirement already satisfied: tenacity>=8.2.3 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (9.1.2)\n",
      "Requirement already satisfied: pyyaml>=6.0.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (6.0.2)\n",
      "Collecting mmh3>=4.0.1 (from chromadb)\n",
      "  Downloading mmh3-5.2.0-cp313-cp313-win_amd64.whl.metadata (14 kB)\n",
      "Requirement already satisfied: orjson>=3.9.12 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (3.10.18)\n",
      "Requirement already satisfied: httpx>=0.27.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (0.28.1)\n",
      "Collecting rich>=10.11.0 (from chromadb)\n",
      "  Downloading rich-14.1.0-py3-none-any.whl.metadata (18 kB)\n",
      "Requirement already satisfied: jsonschema>=4.19.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from chromadb) (4.23.0)\n",
      "Requirement already satisfied: requests<3.0,>=2.7 in a:\\work\\langchain\\lang\\lib\\site-packages (from posthog<6.0.0,>=2.4.0->chromadb) (2.32.3)\n",
      "Requirement already satisfied: six>=1.5 in a:\\work\\langchain\\lang\\lib\\site-packages (from posthog<6.0.0,>=2.4.0->chromadb) (1.17.0)\n",
      "Requirement already satisfied: python-dateutil>=2.2 in a:\\work\\langchain\\lang\\lib\\site-packages (from posthog<6.0.0,>=2.4.0->chromadb) (2.9.0.post0)\n",
      "Collecting backoff>=1.10.0 (from posthog<6.0.0,>=2.4.0->chromadb)\n",
      "  Downloading backoff-2.2.1-py3-none-any.whl.metadata (14 kB)\n",
      "Requirement already satisfied: distro>=1.5.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from posthog<6.0.0,>=2.4.0->chromadb) (1.9.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in a:\\work\\langchain\\lang\\lib\\site-packages (from requests<3.0,>=2.7->posthog<6.0.0,>=2.4.0->chromadb) (3.4.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in a:\\work\\langchain\\lang\\lib\\site-packages (from requests<3.0,>=2.7->posthog<6.0.0,>=2.4.0->chromadb) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in a:\\work\\langchain\\lang\\lib\\site-packages (from requests<3.0,>=2.7->posthog<6.0.0,>=2.4.0->chromadb) (2.4.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in a:\\work\\langchain\\lang\\lib\\site-packages (from requests<3.0,>=2.7->posthog<6.0.0,>=2.4.0->chromadb) (2025.4.26)\n",
      "Requirement already satisfied: packaging>=19.1 in a:\\work\\langchain\\lang\\lib\\site-packages (from build>=1.0.3->chromadb) (24.2)\n",
      "Collecting pyproject_hooks (from build>=1.0.3->chromadb)\n",
      "  Downloading pyproject_hooks-1.2.0-py3-none-any.whl.metadata (1.3 kB)\n",
      "Requirement already satisfied: colorama in a:\\work\\langchain\\lang\\lib\\site-packages (from build>=1.0.3->chromadb) (0.4.6)\n",
      "Requirement already satisfied: anyio in a:\\work\\langchain\\lang\\lib\\site-packages (from httpx>=0.27.0->chromadb) (4.9.0)\n",
      "Requirement already satisfied: httpcore==1.* in a:\\work\\langchain\\lang\\lib\\site-packages (from httpx>=0.27.0->chromadb) (1.0.9)\n",
      "Requirement already satisfied: h11>=0.16 in a:\\work\\langchain\\lang\\lib\\site-packages (from httpcore==1.*->httpx>=0.27.0->chromadb) (0.16.0)\n",
      "Requirement already satisfied: attrs>=22.2.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from jsonschema>=4.19.0->chromadb) (25.3.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in a:\\work\\langchain\\lang\\lib\\site-packages (from jsonschema>=4.19.0->chromadb) (2025.4.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in a:\\work\\langchain\\lang\\lib\\site-packages (from jsonschema>=4.19.0->chromadb) (0.36.2)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in a:\\work\\langchain\\lang\\lib\\site-packages (from jsonschema>=4.19.0->chromadb) (0.24.0)\n",
      "Requirement already satisfied: google-auth>=1.0.1 in a:\\work\\langchain\\lang\\lib\\site-packages (from kubernetes>=28.1.0->chromadb) (2.40.1)\n",
      "Collecting websocket-client!=0.40.0,!=0.41.*,!=0.42.*,>=0.32.0 (from kubernetes>=28.1.0->chromadb)\n",
      "  Using cached websocket_client-1.8.0-py3-none-any.whl.metadata (8.0 kB)\n",
      "Collecting requests-oauthlib (from kubernetes>=28.1.0->chromadb)\n",
      "  Downloading requests_oauthlib-2.0.0-py2.py3-none-any.whl.metadata (11 kB)\n",
      "Collecting oauthlib>=3.2.2 (from kubernetes>=28.1.0->chromadb)\n",
      "  Downloading oauthlib-3.3.1-py3-none-any.whl.metadata (7.9 kB)\n",
      "Collecting durationpy>=0.7 (from kubernetes>=28.1.0->chromadb)\n",
      "  Downloading durationpy-0.10-py3-none-any.whl.metadata (340 bytes)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from google-auth>=1.0.1->kubernetes>=28.1.0->chromadb) (5.5.2)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in a:\\work\\langchain\\lang\\lib\\site-packages (from google-auth>=1.0.1->kubernetes>=28.1.0->chromadb) (0.4.2)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in a:\\work\\langchain\\lang\\lib\\site-packages (from google-auth>=1.0.1->kubernetes>=28.1.0->chromadb) (4.9.1)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in a:\\work\\langchain\\lang\\lib\\site-packages (from rsa<5,>=3.1.4->google-auth>=1.0.1->kubernetes>=28.1.0->chromadb) (0.6.1)\n",
      "Collecting coloredlogs (from onnxruntime>=1.14.1->chromadb)\n",
      "  Downloading coloredlogs-15.0.1-py2.py3-none-any.whl.metadata (12 kB)\n",
      "Collecting flatbuffers (from onnxruntime>=1.14.1->chromadb)\n",
      "  Downloading flatbuffers-25.2.10-py2.py3-none-any.whl.metadata (875 bytes)\n",
      "Requirement already satisfied: protobuf in a:\\work\\langchain\\lang\\lib\\site-packages (from onnxruntime>=1.14.1->chromadb) (5.29.4)\n",
      "Requirement already satisfied: sympy in a:\\work\\langchain\\lang\\lib\\site-packages (from onnxruntime>=1.14.1->chromadb) (1.14.0)\n",
      "Collecting importlib-metadata<8.8.0,>=6.0 (from opentelemetry-api>=1.2.0->chromadb)\n",
      "  Downloading importlib_metadata-8.7.0-py3-none-any.whl.metadata (4.8 kB)\n",
      "Collecting zipp>=3.20 (from importlib-metadata<8.8.0,>=6.0->opentelemetry-api>=1.2.0->chromadb)\n",
      "  Downloading zipp-3.23.0-py3-none-any.whl.metadata (3.6 kB)\n",
      "Requirement already satisfied: googleapis-common-protos~=1.57 in a:\\work\\langchain\\lang\\lib\\site-packages (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb) (1.70.0)\n",
      "Collecting opentelemetry-exporter-otlp-proto-common==1.36.0 (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb)\n",
      "  Downloading opentelemetry_exporter_otlp_proto_common-1.36.0-py3-none-any.whl.metadata (1.8 kB)\n",
      "Collecting opentelemetry-proto==1.36.0 (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb)\n",
      "  Downloading opentelemetry_proto-1.36.0-py3-none-any.whl.metadata (2.3 kB)\n",
      "Collecting opentelemetry-semantic-conventions==0.57b0 (from opentelemetry-sdk>=1.2.0->chromadb)\n",
      "  Downloading opentelemetry_semantic_conventions-0.57b0-py3-none-any.whl.metadata (2.4 kB)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from pydantic>=1.9->chromadb) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.33.2 in a:\\work\\langchain\\lang\\lib\\site-packages (from pydantic>=1.9->chromadb) (2.33.2)\n",
      "Requirement already satisfied: typing-inspection>=0.4.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from pydantic>=1.9->chromadb) (0.4.0)\n",
      "Collecting markdown-it-py>=2.2.0 (from rich>=10.11.0->chromadb)\n",
      "  Downloading markdown_it_py-4.0.0-py3-none-any.whl.metadata (7.3 kB)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from rich>=10.11.0->chromadb) (2.19.1)\n",
      "Collecting mdurl~=0.1 (from markdown-it-py>=2.2.0->rich>=10.11.0->chromadb)\n",
      "  Downloading mdurl-0.1.2-py3-none-any.whl.metadata (1.6 kB)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.16.4 in a:\\work\\langchain\\lang\\lib\\site-packages (from tokenizers>=0.13.2->chromadb) (0.34.4)\n",
      "Requirement already satisfied: filelock in a:\\work\\langchain\\lang\\lib\\site-packages (from huggingface-hub<1.0,>=0.16.4->tokenizers>=0.13.2->chromadb) (3.18.0)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from huggingface-hub<1.0,>=0.16.4->tokenizers>=0.13.2->chromadb) (2025.3.2)\n",
      "Requirement already satisfied: click>=8.0.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from typer>=0.9.0->chromadb) (8.2.0)\n",
      "Collecting shellingham>=1.3.0 (from typer>=0.9.0->chromadb)\n",
      "  Downloading shellingham-1.5.4-py2.py3-none-any.whl.metadata (3.5 kB)\n",
      "Collecting httptools>=0.6.3 (from uvicorn[standard]>=0.18.3->chromadb)\n",
      "  Downloading httptools-0.6.4-cp313-cp313-win_amd64.whl.metadata (3.7 kB)\n",
      "Requirement already satisfied: python-dotenv>=0.13 in a:\\work\\langchain\\lang\\lib\\site-packages (from uvicorn[standard]>=0.18.3->chromadb) (1.1.0)\n",
      "Collecting watchfiles>=0.13 (from uvicorn[standard]>=0.18.3->chromadb)\n",
      "  Downloading watchfiles-1.1.0-cp313-cp313-win_amd64.whl.metadata (5.0 kB)\n",
      "Collecting websockets>=10.4 (from uvicorn[standard]>=0.18.3->chromadb)\n",
      "  Downloading websockets-15.0.1-cp313-cp313-win_amd64.whl.metadata (7.0 kB)\n",
      "Requirement already satisfied: sniffio>=1.1 in a:\\work\\langchain\\lang\\lib\\site-packages (from anyio->httpx>=0.27.0->chromadb) (1.3.1)\n",
      "Collecting humanfriendly>=9.1 (from coloredlogs->onnxruntime>=1.14.1->chromadb)\n",
      "  Downloading humanfriendly-10.0-py2.py3-none-any.whl.metadata (9.2 kB)\n",
      "Collecting pyreadline3 (from humanfriendly>=9.1->coloredlogs->onnxruntime>=1.14.1->chromadb)\n",
      "  Downloading pyreadline3-3.5.4-py3-none-any.whl.metadata (4.7 kB)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in a:\\work\\langchain\\lang\\lib\\site-packages (from sympy->onnxruntime>=1.14.1->chromadb) (1.3.0)\n",
      "Downloading chromadb-1.0.17-cp39-abi3-win_amd64.whl (19.8 MB)\n",
      "   ---------------------------------------- 0.0/19.8 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.3/19.8 MB ? eta -:--:--\n",
      "   - -------------------------------------- 0.5/19.8 MB 1.8 MB/s eta 0:00:11\n",
      "   - -------------------------------------- 0.8/19.8 MB 1.8 MB/s eta 0:00:11\n",
      "   -- ------------------------------------- 1.3/19.8 MB 1.7 MB/s eta 0:00:11\n",
      "   --- ------------------------------------ 1.6/19.8 MB 1.7 MB/s eta 0:00:11\n",
      "   ---- ----------------------------------- 2.1/19.8 MB 1.8 MB/s eta 0:00:10\n",
      "   ---- ----------------------------------- 2.1/19.8 MB 1.8 MB/s eta 0:00:10\n",
      "   ---- ----------------------------------- 2.4/19.8 MB 1.5 MB/s eta 0:00:12\n",
      "   ----- ---------------------------------- 2.9/19.8 MB 1.6 MB/s eta 0:00:11\n",
      "   ------ --------------------------------- 3.1/19.8 MB 1.6 MB/s eta 0:00:11\n",
      "   ------- -------------------------------- 3.7/19.8 MB 1.7 MB/s eta 0:00:10\n",
      "   ------- -------------------------------- 3.9/19.8 MB 1.6 MB/s eta 0:00:10\n",
      "   -------- ------------------------------- 4.2/19.8 MB 1.6 MB/s eta 0:00:10\n",
      "   -------- ------------------------------- 4.2/19.8 MB 1.6 MB/s eta 0:00:10\n",
      "   --------- ------------------------------ 4.7/19.8 MB 1.5 MB/s eta 0:00:10\n",
      "   ---------- ----------------------------- 5.0/19.8 MB 1.6 MB/s eta 0:00:10\n",
      "   ---------- ----------------------------- 5.0/19.8 MB 1.6 MB/s eta 0:00:10\n",
      "   ---------- ----------------------------- 5.0/19.8 MB 1.6 MB/s eta 0:00:10\n",
      "   ----------- ---------------------------- 5.5/19.8 MB 1.4 MB/s eta 0:00:11\n",
      "   ------------ --------------------------- 6.0/19.8 MB 1.5 MB/s eta 0:00:10\n",
      "   ------------ --------------------------- 6.3/19.8 MB 1.5 MB/s eta 0:00:10\n",
      "   ------------- -------------------------- 6.6/19.8 MB 1.4 MB/s eta 0:00:10\n",
      "   ------------- -------------------------- 6.8/19.8 MB 1.4 MB/s eta 0:00:09\n",
      "   -------------- ------------------------- 7.1/19.8 MB 1.5 MB/s eta 0:00:09\n",
      "   -------------- ------------------------- 7.3/19.8 MB 1.4 MB/s eta 0:00:09\n",
      "   --------------- ------------------------ 7.6/19.8 MB 1.4 MB/s eta 0:00:09\n",
      "   --------------- ------------------------ 7.9/19.8 MB 1.5 MB/s eta 0:00:09\n",
      "   ---------------- ----------------------- 8.4/19.8 MB 1.5 MB/s eta 0:00:08\n",
      "   ----------------- ---------------------- 8.7/19.8 MB 1.5 MB/s eta 0:00:08\n",
      "   ------------------ --------------------- 8.9/19.8 MB 1.5 MB/s eta 0:00:08\n",
      "   ------------------ --------------------- 8.9/19.8 MB 1.5 MB/s eta 0:00:08\n",
      "   ------------------ --------------------- 9.2/19.8 MB 1.4 MB/s eta 0:00:08\n",
      "   ------------------ --------------------- 9.2/19.8 MB 1.4 MB/s eta 0:00:08\n",
      "   ------------------- -------------------- 9.4/19.8 MB 1.4 MB/s eta 0:00:08\n",
      "   ------------------- -------------------- 9.7/19.8 MB 1.4 MB/s eta 0:00:08\n",
      "   ------------------- -------------------- 9.7/19.8 MB 1.4 MB/s eta 0:00:08\n",
      "   ------------------- -------------------- 9.7/19.8 MB 1.4 MB/s eta 0:00:08\n",
      "   -------------------- ------------------- 10.0/19.8 MB 1.3 MB/s eta 0:00:08\n",
      "   -------------------- ------------------- 10.0/19.8 MB 1.3 MB/s eta 0:00:08\n",
      "   -------------------- ------------------- 10.2/19.8 MB 1.3 MB/s eta 0:00:08\n",
      "   --------------------- ------------------ 10.5/19.8 MB 1.2 MB/s eta 0:00:08\n",
      "   --------------------- ------------------ 10.5/19.8 MB 1.2 MB/s eta 0:00:08\n",
      "   --------------------- ------------------ 10.7/19.8 MB 1.2 MB/s eta 0:00:08\n",
      "   ---------------------- ----------------- 11.0/19.8 MB 1.2 MB/s eta 0:00:08\n",
      "   ---------------------- ----------------- 11.3/19.8 MB 1.2 MB/s eta 0:00:07\n",
      "   ---------------------- ----------------- 11.3/19.8 MB 1.2 MB/s eta 0:00:07\n",
      "   ----------------------- ---------------- 11.5/19.8 MB 1.2 MB/s eta 0:00:07\n",
      "   ----------------------- ---------------- 11.5/19.8 MB 1.2 MB/s eta 0:00:07\n",
      "   ----------------------- ---------------- 11.8/19.8 MB 1.2 MB/s eta 0:00:07\n",
      "   ------------------------ --------------- 12.1/19.8 MB 1.2 MB/s eta 0:00:07\n",
      "   ------------------------ --------------- 12.3/19.8 MB 1.2 MB/s eta 0:00:07\n",
      "   ------------------------- -------------- 12.6/19.8 MB 1.2 MB/s eta 0:00:07\n",
      "   ------------------------- -------------- 12.8/19.8 MB 1.2 MB/s eta 0:00:06\n",
      "   ------------------------- -------------- 12.8/19.8 MB 1.2 MB/s eta 0:00:06\n",
      "   -------------------------- ------------- 13.1/19.8 MB 1.2 MB/s eta 0:00:06\n",
      "   --------------------------- ------------ 13.4/19.8 MB 1.2 MB/s eta 0:00:06\n",
      "   --------------------------- ------------ 13.6/19.8 MB 1.2 MB/s eta 0:00:06\n",
      "   --------------------------- ------------ 13.6/19.8 MB 1.2 MB/s eta 0:00:06\n",
      "   ---------------------------- ----------- 13.9/19.8 MB 1.2 MB/s eta 0:00:06\n",
      "   ---------------------------- ----------- 14.2/19.8 MB 1.2 MB/s eta 0:00:05\n",
      "   ----------------------------- ---------- 14.4/19.8 MB 1.1 MB/s eta 0:00:05\n",
      "   ----------------------------- ---------- 14.7/19.8 MB 1.1 MB/s eta 0:00:05\n",
      "   ------------------------------ --------- 14.9/19.8 MB 1.2 MB/s eta 0:00:05\n",
      "   ------------------------------ --------- 15.2/19.8 MB 1.1 MB/s eta 0:00:04\n",
      "   ------------------------------- -------- 15.5/19.8 MB 1.2 MB/s eta 0:00:04\n",
      "   ------------------------------- -------- 15.5/19.8 MB 1.2 MB/s eta 0:00:04\n",
      "   ------------------------------- -------- 15.7/19.8 MB 1.1 MB/s eta 0:00:04\n",
      "   -------------------------------- ------- 16.0/19.8 MB 1.1 MB/s eta 0:00:04\n",
      "   -------------------------------- ------- 16.3/19.8 MB 1.2 MB/s eta 0:00:04\n",
      "   --------------------------------- ------ 16.5/19.8 MB 1.2 MB/s eta 0:00:03\n",
      "   --------------------------------- ------ 16.8/19.8 MB 1.2 MB/s eta 0:00:03\n",
      "   ----------------------------------- ---- 17.3/19.8 MB 1.2 MB/s eta 0:00:03\n",
      "   ----------------------------------- ---- 17.6/19.8 MB 1.2 MB/s eta 0:00:02\n",
      "   ----------------------------------- ---- 17.6/19.8 MB 1.2 MB/s eta 0:00:02\n",
      "   ------------------------------------ --- 17.8/19.8 MB 1.2 MB/s eta 0:00:02\n",
      "   ------------------------------------ --- 18.1/19.8 MB 1.2 MB/s eta 0:00:02\n",
      "   ------------------------------------- -- 18.4/19.8 MB 1.2 MB/s eta 0:00:02\n",
      "   ------------------------------------- -- 18.4/19.8 MB 1.2 MB/s eta 0:00:02\n",
      "   ------------------------------------- -- 18.6/19.8 MB 1.2 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 18.6/19.8 MB 1.2 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 18.9/19.8 MB 1.1 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 18.9/19.8 MB 1.1 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 19.1/19.8 MB 1.1 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 19.1/19.8 MB 1.1 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 19.1/19.8 MB 1.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------  19.4/19.8 MB 1.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------  19.4/19.8 MB 1.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------  19.7/19.8 MB 1.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 19.8/19.8 MB 1.1 MB/s eta 0:00:00\n",
      "Downloading posthog-5.4.0-py3-none-any.whl (105 kB)\n",
      "Downloading backoff-2.2.1-py3-none-any.whl (15 kB)\n",
      "Downloading bcrypt-4.3.0-cp39-abi3-win_amd64.whl (152 kB)\n",
      "Downloading build-1.3.0-py3-none-any.whl (23 kB)\n",
      "Downloading kubernetes-33.1.0-py2.py3-none-any.whl (1.9 MB)\n",
      "   ---------------------------------------- 0.0/1.9 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/1.9 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/1.9 MB ? eta -:--:--\n",
      "   ----- ---------------------------------- 0.3/1.9 MB ? eta -:--:--\n",
      "   ----- ---------------------------------- 0.3/1.9 MB ? eta -:--:--\n",
      "   ---------- ----------------------------- 0.5/1.9 MB 555.1 kB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 0.5/1.9 MB 555.1 kB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 0.8/1.9 MB 614.2 kB/s eta 0:00:02\n",
      "   --------------------- ------------------ 1.0/1.9 MB 641.0 kB/s eta 0:00:02\n",
      "   --------------------- ------------------ 1.0/1.9 MB 641.0 kB/s eta 0:00:02\n",
      "   --------------------- ------------------ 1.0/1.9 MB 641.0 kB/s eta 0:00:02\n",
      "   --------------------------- ------------ 1.3/1.9 MB 631.0 kB/s eta 0:00:01\n",
      "   --------------------------- ------------ 1.3/1.9 MB 631.0 kB/s eta 0:00:01\n",
      "   --------------------------- ------------ 1.3/1.9 MB 631.0 kB/s eta 0:00:01\n",
      "   --------------------------- ------------ 1.3/1.9 MB 631.0 kB/s eta 0:00:01\n",
      "   --------------------------- ------------ 1.3/1.9 MB 631.0 kB/s eta 0:00:01\n",
      "   -------------------------------- ------- 1.6/1.9 MB 459.3 kB/s eta 0:00:01\n",
      "   -------------------------------- ------- 1.6/1.9 MB 459.3 kB/s eta 0:00:01\n",
      "   -------------------------------- ------- 1.6/1.9 MB 459.3 kB/s eta 0:00:01\n",
      "   -------------------------------- ------- 1.6/1.9 MB 459.3 kB/s eta 0:00:01\n",
      "   ------------------------------------- -- 1.8/1.9 MB 439.2 kB/s eta 0:00:01\n",
      "   ---------------------------------------- 1.9/1.9 MB 442.3 kB/s eta 0:00:00\n",
      "Downloading durationpy-0.10-py3-none-any.whl (3.9 kB)\n",
      "Downloading mmh3-5.2.0-cp313-cp313-win_amd64.whl (41 kB)\n",
      "Downloading oauthlib-3.3.1-py3-none-any.whl (160 kB)\n",
      "Downloading onnxruntime-1.22.1-cp313-cp313-win_amd64.whl (12.7 MB)\n",
      "   ---------------------------------------- 0.0/12.7 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/12.7 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/12.7 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.3/12.7 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.3/12.7 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.3/12.7 MB ? eta -:--:--\n",
      "   - -------------------------------------- 0.5/12.7 MB 506.9 kB/s eta 0:00:25\n",
      "   - -------------------------------------- 0.5/12.7 MB 506.9 kB/s eta 0:00:25\n",
      "   -- ------------------------------------- 0.8/12.7 MB 526.1 kB/s eta 0:00:23\n",
      "   -- ------------------------------------- 0.8/12.7 MB 526.1 kB/s eta 0:00:23\n",
      "   -- ------------------------------------- 0.8/12.7 MB 526.1 kB/s eta 0:00:23\n",
      "   --- ------------------------------------ 1.0/12.7 MB 516.0 kB/s eta 0:00:23\n",
      "   --- ------------------------------------ 1.0/12.7 MB 516.0 kB/s eta 0:00:23\n",
      "   ---- ----------------------------------- 1.3/12.7 MB 515.6 kB/s eta 0:00:23\n",
      "   ---- ----------------------------------- 1.3/12.7 MB 515.6 kB/s eta 0:00:23\n",
      "   ---- ----------------------------------- 1.6/12.7 MB 529.9 kB/s eta 0:00:21\n",
      "   ---- ----------------------------------- 1.6/12.7 MB 529.9 kB/s eta 0:00:21\n",
      "   ----- ---------------------------------- 1.8/12.7 MB 547.9 kB/s eta 0:00:20\n",
      "   ----- ---------------------------------- 1.8/12.7 MB 547.9 kB/s eta 0:00:20\n",
      "   ------ --------------------------------- 2.1/12.7 MB 561.9 kB/s eta 0:00:19\n",
      "   ------ --------------------------------- 2.1/12.7 MB 561.9 kB/s eta 0:00:19\n",
      "   ------- -------------------------------- 2.4/12.7 MB 571.1 kB/s eta 0:00:19\n",
      "   ------- -------------------------------- 2.4/12.7 MB 571.1 kB/s eta 0:00:19\n",
      "   ------- -------------------------------- 2.4/12.7 MB 571.1 kB/s eta 0:00:19\n",
      "   -------- ------------------------------- 2.6/12.7 MB 545.7 kB/s eta 0:00:19\n",
      "   -------- ------------------------------- 2.6/12.7 MB 545.7 kB/s eta 0:00:19\n",
      "   -------- ------------------------------- 2.6/12.7 MB 545.7 kB/s eta 0:00:19\n",
      "   -------- ------------------------------- 2.6/12.7 MB 545.7 kB/s eta 0:00:19\n",
      "   --------- ------------------------------ 2.9/12.7 MB 514.7 kB/s eta 0:00:20\n",
      "   --------- ------------------------------ 2.9/12.7 MB 514.7 kB/s eta 0:00:20\n",
      "   --------- ------------------------------ 2.9/12.7 MB 514.7 kB/s eta 0:00:20\n",
      "   --------- ------------------------------ 3.1/12.7 MB 506.6 kB/s eta 0:00:19\n",
      "   --------- ------------------------------ 3.1/12.7 MB 506.6 kB/s eta 0:00:19\n",
      "   --------- ------------------------------ 3.1/12.7 MB 506.6 kB/s eta 0:00:19\n",
      "   --------- ------------------------------ 3.1/12.7 MB 506.6 kB/s eta 0:00:19\n",
      "   ---------- ----------------------------- 3.4/12.7 MB 481.7 kB/s eta 0:00:20\n",
      "   ---------- ----------------------------- 3.4/12.7 MB 481.7 kB/s eta 0:00:20\n",
      "   ---------- ----------------------------- 3.4/12.7 MB 481.7 kB/s eta 0:00:20\n",
      "   ----------- ---------------------------- 3.7/12.7 MB 481.0 kB/s eta 0:00:19\n",
      "   ----------- ---------------------------- 3.7/12.7 MB 481.0 kB/s eta 0:00:19\n",
      "   ------------ --------------------------- 3.9/12.7 MB 485.6 kB/s eta 0:00:19\n",
      "   ------------ --------------------------- 3.9/12.7 MB 485.6 kB/s eta 0:00:19\n",
      "   ------------ --------------------------- 3.9/12.7 MB 485.6 kB/s eta 0:00:19\n",
      "   ------------ --------------------------- 3.9/12.7 MB 485.6 kB/s eta 0:00:19\n",
      "   ------------ --------------------------- 3.9/12.7 MB 485.6 kB/s eta 0:00:19\n",
      "   ------------ --------------------------- 3.9/12.7 MB 485.6 kB/s eta 0:00:19\n",
      "   ------------ --------------------------- 3.9/12.7 MB 485.6 kB/s eta 0:00:19\n",
      "   ------------- -------------------------- 4.2/12.7 MB 433.7 kB/s eta 0:00:20\n",
      "   ------------- -------------------------- 4.2/12.7 MB 433.7 kB/s eta 0:00:20\n",
      "   ------------- -------------------------- 4.2/12.7 MB 433.7 kB/s eta 0:00:20\n",
      "   ------------- -------------------------- 4.2/12.7 MB 433.7 kB/s eta 0:00:20\n",
      "   ------------- -------------------------- 4.2/12.7 MB 433.7 kB/s eta 0:00:20\n",
      "   -------------- ------------------------- 4.5/12.7 MB 421.4 kB/s eta 0:00:20\n",
      "   -------------- ------------------------- 4.5/12.7 MB 421.4 kB/s eta 0:00:20\n",
      "   -------------- ------------------------- 4.7/12.7 MB 425.3 kB/s eta 0:00:19\n",
      "   -------------- ------------------------- 4.7/12.7 MB 425.3 kB/s eta 0:00:19\n",
      "   --------------- ------------------------ 5.0/12.7 MB 430.7 kB/s eta 0:00:18\n",
      "   --------------- ------------------------ 5.0/12.7 MB 430.7 kB/s eta 0:00:18\n",
      "   --------------- ------------------------ 5.0/12.7 MB 430.7 kB/s eta 0:00:18\n",
      "   ---------------- ----------------------- 5.2/12.7 MB 432.4 kB/s eta 0:00:18\n",
      "   ---------------- ----------------------- 5.2/12.7 MB 432.4 kB/s eta 0:00:18\n",
      "   ---------------- ----------------------- 5.2/12.7 MB 432.4 kB/s eta 0:00:18\n",
      "   ----------------- ---------------------- 5.5/12.7 MB 430.0 kB/s eta 0:00:17\n",
      "   ----------------- ---------------------- 5.5/12.7 MB 430.0 kB/s eta 0:00:17\n",
      "   ----------------- ---------------------- 5.5/12.7 MB 430.0 kB/s eta 0:00:17\n",
      "   ------------------ --------------------- 5.8/12.7 MB 431.5 kB/s eta 0:00:17\n",
      "   ------------------ --------------------- 5.8/12.7 MB 431.5 kB/s eta 0:00:17\n",
      "   ------------------ --------------------- 5.8/12.7 MB 431.5 kB/s eta 0:00:17\n",
      "   ------------------ --------------------- 5.8/12.7 MB 431.5 kB/s eta 0:00:17\n",
      "   ------------------ --------------------- 5.8/12.7 MB 431.5 kB/s eta 0:00:17\n",
      "   ------------------- -------------------- 6.0/12.7 MB 419.7 kB/s eta 0:00:16\n",
      "   ------------------- -------------------- 6.0/12.7 MB 419.7 kB/s eta 0:00:16\n",
      "   ------------------- -------------------- 6.0/12.7 MB 419.7 kB/s eta 0:00:16\n",
      "   ------------------- -------------------- 6.0/12.7 MB 419.7 kB/s eta 0:00:16\n",
      "   ------------------- -------------------- 6.0/12.7 MB 419.7 kB/s eta 0:00:16\n",
      "   ------------------- -------------------- 6.3/12.7 MB 408.7 kB/s eta 0:00:16\n",
      "   ------------------- -------------------- 6.3/12.7 MB 408.7 kB/s eta 0:00:16\n",
      "   ------------------- -------------------- 6.3/12.7 MB 408.7 kB/s eta 0:00:16\n",
      "   ------------------- -------------------- 6.3/12.7 MB 408.7 kB/s eta 0:00:16\n",
      "   -------------------- ------------------- 6.6/12.7 MB 400.6 kB/s eta 0:00:16\n",
      "   -------------------- ------------------- 6.6/12.7 MB 400.6 kB/s eta 0:00:16\n",
      "   -------------------- ------------------- 6.6/12.7 MB 400.6 kB/s eta 0:00:16\n",
      "   -------------------- ------------------- 6.6/12.7 MB 400.6 kB/s eta 0:00:16\n",
      "   --------------------- ------------------ 6.8/12.7 MB 397.0 kB/s eta 0:00:15\n",
      "   --------------------- ------------------ 6.8/12.7 MB 397.0 kB/s eta 0:00:15\n",
      "   --------------------- ------------------ 6.8/12.7 MB 397.0 kB/s eta 0:00:15\n",
      "   ---------------------- ----------------- 7.1/12.7 MB 397.6 kB/s eta 0:00:15\n",
      "   ---------------------- ----------------- 7.1/12.7 MB 397.6 kB/s eta 0:00:15\n",
      "   ---------------------- ----------------- 7.1/12.7 MB 397.6 kB/s eta 0:00:15\n",
      "   ----------------------- ---------------- 7.3/12.7 MB 398.7 kB/s eta 0:00:14\n",
      "   ----------------------- ---------------- 7.3/12.7 MB 398.7 kB/s eta 0:00:14\n",
      "   ----------------------- ---------------- 7.3/12.7 MB 398.7 kB/s eta 0:00:14\n",
      "   ----------------------- ---------------- 7.6/12.7 MB 400.1 kB/s eta 0:00:13\n",
      "   ----------------------- ---------------- 7.6/12.7 MB 400.1 kB/s eta 0:00:13\n",
      "   ----------------------- ---------------- 7.6/12.7 MB 400.1 kB/s eta 0:00:13\n",
      "   ----------------------- ---------------- 7.6/12.7 MB 400.1 kB/s eta 0:00:13\n",
      "   ------------------------ --------------- 7.9/12.7 MB 399.7 kB/s eta 0:00:13\n",
      "   ------------------------ --------------- 7.9/12.7 MB 399.7 kB/s eta 0:00:13\n",
      "   ------------------------ --------------- 7.9/12.7 MB 399.7 kB/s eta 0:00:13\n",
      "   ------------------------ --------------- 7.9/12.7 MB 399.7 kB/s eta 0:00:13\n",
      "   ------------------------ --------------- 7.9/12.7 MB 399.7 kB/s eta 0:00:13\n",
      "   ------------------------- -------------- 8.1/12.7 MB 391.9 kB/s eta 0:00:12\n",
      "   ------------------------- -------------- 8.1/12.7 MB 391.9 kB/s eta 0:00:12\n",
      "   -------------------------- ------------- 8.4/12.7 MB 394.5 kB/s eta 0:00:11\n",
      "   -------------------------- ------------- 8.4/12.7 MB 394.5 kB/s eta 0:00:11\n",
      "   -------------------------- ------------- 8.4/12.7 MB 394.5 kB/s eta 0:00:11\n",
      "   --------------------------- ------------ 8.7/12.7 MB 397.4 kB/s eta 0:00:11\n",
      "   --------------------------- ------------ 8.7/12.7 MB 397.4 kB/s eta 0:00:11\n",
      "   --------------------------- ------------ 8.7/12.7 MB 397.4 kB/s eta 0:00:11\n",
      "   --------------------------- ------------ 8.7/12.7 MB 397.4 kB/s eta 0:00:11\n",
      "   ---------------------------- ----------- 8.9/12.7 MB 393.8 kB/s eta 0:00:10\n",
      "   ---------------------------- ----------- 8.9/12.7 MB 393.8 kB/s eta 0:00:10\n",
      "   ---------------------------- ----------- 8.9/12.7 MB 393.8 kB/s eta 0:00:10\n",
      "   ---------------------------- ----------- 9.2/12.7 MB 395.0 kB/s eta 0:00:09\n",
      "   ---------------------------- ----------- 9.2/12.7 MB 395.0 kB/s eta 0:00:09\n",
      "   ----------------------------- ---------- 9.4/12.7 MB 398.0 kB/s eta 0:00:09\n",
      "   ----------------------------- ---------- 9.4/12.7 MB 398.0 kB/s eta 0:00:09\n",
      "   ----------------------------- ---------- 9.4/12.7 MB 398.0 kB/s eta 0:00:09\n",
      "   ------------------------------ --------- 9.7/12.7 MB 400.4 kB/s eta 0:00:08\n",
      "   ------------------------------ --------- 9.7/12.7 MB 400.4 kB/s eta 0:00:08\n",
      "   ------------------------------- -------- 10.0/12.7 MB 403.3 kB/s eta 0:00:07\n",
      "   ------------------------------- -------- 10.0/12.7 MB 403.3 kB/s eta 0:00:07\n",
      "   -------------------------------- ------- 10.2/12.7 MB 405.6 kB/s eta 0:00:07\n",
      "   -------------------------------- ------- 10.2/12.7 MB 405.6 kB/s eta 0:00:07\n",
      "   -------------------------------- ------- 10.2/12.7 MB 405.6 kB/s eta 0:00:07\n",
      "   --------------------------------- ------ 10.5/12.7 MB 407.0 kB/s eta 0:00:06\n",
      "   --------------------------------- ------ 10.5/12.7 MB 407.0 kB/s eta 0:00:06\n",
      "   --------------------------------- ------ 10.5/12.7 MB 407.0 kB/s eta 0:00:06\n",
      "   --------------------------------- ------ 10.7/12.7 MB 406.5 kB/s eta 0:00:05\n",
      "   --------------------------------- ------ 10.7/12.7 MB 406.5 kB/s eta 0:00:05\n",
      "   ---------------------------------- ----- 11.0/12.7 MB 410.9 kB/s eta 0:00:05\n",
      "   ---------------------------------- ----- 11.0/12.7 MB 410.9 kB/s eta 0:00:05\n",
      "   ---------------------------------- ----- 11.0/12.7 MB 410.9 kB/s eta 0:00:05\n",
      "   ----------------------------------- ---- 11.3/12.7 MB 413.2 kB/s eta 0:00:04\n",
      "   ----------------------------------- ---- 11.3/12.7 MB 413.2 kB/s eta 0:00:04\n",
      "   ----------------------------------- ---- 11.3/12.7 MB 413.2 kB/s eta 0:00:04\n",
      "   ----------------------------------- ---- 11.3/12.7 MB 413.2 kB/s eta 0:00:04\n",
      "   ------------------------------------ --- 11.5/12.7 MB 409.2 kB/s eta 0:00:03\n",
      "   ------------------------------------ --- 11.5/12.7 MB 409.2 kB/s eta 0:00:03\n",
      "   ------------------------------------ --- 11.5/12.7 MB 409.2 kB/s eta 0:00:03\n",
      "   ------------------------------------- -- 11.8/12.7 MB 409.5 kB/s eta 0:00:03\n",
      "   ------------------------------------- -- 11.8/12.7 MB 409.5 kB/s eta 0:00:03\n",
      "   -------------------------------------- - 12.1/12.7 MB 412.0 kB/s eta 0:00:02\n",
      "   -------------------------------------- - 12.1/12.7 MB 412.0 kB/s eta 0:00:02\n",
      "   -------------------------------------- - 12.3/12.7 MB 414.6 kB/s eta 0:00:01\n",
      "   -------------------------------------- - 12.3/12.7 MB 414.6 kB/s eta 0:00:01\n",
      "   -------------------------------------- - 12.3/12.7 MB 414.6 kB/s eta 0:00:01\n",
      "   ---------------------------------------  12.6/12.7 MB 416.3 kB/s eta 0:00:01\n",
      "   ---------------------------------------- 12.7/12.7 MB 416.4 kB/s eta 0:00:00\n",
      "Downloading opentelemetry_api-1.36.0-py3-none-any.whl (65 kB)\n",
      "Downloading importlib_metadata-8.7.0-py3-none-any.whl (27 kB)\n",
      "Downloading opentelemetry_exporter_otlp_proto_grpc-1.36.0-py3-none-any.whl (18 kB)\n",
      "Downloading opentelemetry_exporter_otlp_proto_common-1.36.0-py3-none-any.whl (18 kB)\n",
      "Downloading opentelemetry_proto-1.36.0-py3-none-any.whl (72 kB)\n",
      "Downloading opentelemetry_sdk-1.36.0-py3-none-any.whl (119 kB)\n",
      "Downloading opentelemetry_semantic_conventions-0.57b0-py3-none-any.whl (201 kB)\n",
      "Using cached overrides-7.7.0-py3-none-any.whl (17 kB)\n",
      "Downloading pybase64-1.4.2-cp313-cp313-win_amd64.whl (35 kB)\n",
      "Downloading rich-14.1.0-py3-none-any.whl (243 kB)\n",
      "Downloading markdown_it_py-4.0.0-py3-none-any.whl (87 kB)\n",
      "Downloading mdurl-0.1.2-py3-none-any.whl (10.0 kB)\n",
      "Downloading typer-0.16.0-py3-none-any.whl (46 kB)\n",
      "Downloading shellingham-1.5.4-py2.py3-none-any.whl (9.8 kB)\n",
      "Downloading uvicorn-0.35.0-py3-none-any.whl (66 kB)\n",
      "Downloading httptools-0.6.4-cp313-cp313-win_amd64.whl (87 kB)\n",
      "Downloading watchfiles-1.1.0-cp313-cp313-win_amd64.whl (292 kB)\n",
      "Using cached websocket_client-1.8.0-py3-none-any.whl (58 kB)\n",
      "Downloading websockets-15.0.1-cp313-cp313-win_amd64.whl (176 kB)\n",
      "Downloading zipp-3.23.0-py3-none-any.whl (10 kB)\n",
      "Downloading coloredlogs-15.0.1-py2.py3-none-any.whl (46 kB)\n",
      "Downloading humanfriendly-10.0-py2.py3-none-any.whl (86 kB)\n",
      "Downloading flatbuffers-25.2.10-py2.py3-none-any.whl (30 kB)\n",
      "Downloading importlib_resources-6.5.2-py3-none-any.whl (37 kB)\n",
      "Downloading pyproject_hooks-1.2.0-py3-none-any.whl (10 kB)\n",
      "Downloading pyreadline3-3.5.4-py3-none-any.whl (83 kB)\n",
      "Downloading requests_oauthlib-2.0.0-py2.py3-none-any.whl (24 kB)\n",
      "Building wheels for collected packages: pypika\n",
      "  Building wheel for pypika (pyproject.toml): started\n",
      "  Building wheel for pypika (pyproject.toml): finished with status 'done'\n",
      "  Created wheel for pypika: filename=pypika-0.48.9-py2.py3-none-any.whl size=53916 sha256=e675b2c1e76ea6742b819b060064674798a955fc59718bfd94244116dfd74a18\n",
      "  Stored in directory: c:\\users\\log1c\\appdata\\local\\pip\\cache\\wheels\\b4\\f8\\a5\\28e9c1524d320f4b8eefdce0e487b5c2e128dbf2ed1bb4a60b\n",
      "Successfully built pypika\n",
      "Installing collected packages: pypika, flatbuffers, durationpy, zipp, websockets, websocket-client, shellingham, pyreadline3, pyproject_hooks, pybase64, overrides, opentelemetry-proto, oauthlib, mmh3, mdurl, importlib-resources, httptools, bcrypt, backoff, watchfiles, uvicorn, requests-oauthlib, posthog, opentelemetry-exporter-otlp-proto-common, markdown-it-py, importlib-metadata, humanfriendly, build, rich, opentelemetry-api, kubernetes, coloredlogs, typer, opentelemetry-semantic-conventions, onnxruntime, opentelemetry-sdk, opentelemetry-exporter-otlp-proto-grpc, chromadb\n",
      "\n",
      "   - --------------------------------------  1/38 [flatbuffers]\n",
      "   --- ------------------------------------  3/38 [zipp]\n",
      "   ---- -----------------------------------  4/38 [websockets]\n",
      "   ---- -----------------------------------  4/38 [websockets]\n",
      "   ---- -----------------------------------  4/38 [websockets]\n",
      "   ---- -----------------------------------  4/38 [websockets]\n",
      "   ----- ----------------------------------  5/38 [websocket-client]\n",
      "   ----- ----------------------------------  5/38 [websocket-client]\n",
      "   ----- ----------------------------------  5/38 [websocket-client]\n",
      "   ------- --------------------------------  7/38 [pyreadline3]\n",
      "   ------- --------------------------------  7/38 [pyreadline3]\n",
      "   ------- --------------------------------  7/38 [pyreadline3]\n",
      "   ------- --------------------------------  7/38 [pyreadline3]\n",
      "   --------- ------------------------------  9/38 [pybase64]\n",
      "   ---------- ----------------------------- 10/38 [overrides]\n",
      "   ----------- ---------------------------- 11/38 [opentelemetry-proto]\n",
      "   ----------- ---------------------------- 11/38 [opentelemetry-proto]\n",
      "   ------------ --------------------------- 12/38 [oauthlib]\n",
      "   ------------ --------------------------- 12/38 [oauthlib]\n",
      "   ------------ --------------------------- 12/38 [oauthlib]\n",
      "   ------------ --------------------------- 12/38 [oauthlib]\n",
      "   ------------ --------------------------- 12/38 [oauthlib]\n",
      "   ------------ --------------------------- 12/38 [oauthlib]\n",
      "   --------------- ------------------------ 15/38 [importlib-resources]\n",
      "   --------------- ------------------------ 15/38 [importlib-resources]\n",
      "   --------------- ------------------------ 15/38 [importlib-resources]\n",
      "   --------------- ------------------------ 15/38 [importlib-resources]\n",
      "   ------------------ --------------------- 18/38 [backoff]\n",
      "   -------------------- ------------------- 19/38 [watchfiles]\n",
      "   --------------------- ------------------ 20/38 [uvicorn]\n",
      "   --------------------- ------------------ 20/38 [uvicorn]\n",
      "   --------------------- ------------------ 20/38 [uvicorn]\n",
      "   ---------------------- ----------------- 21/38 [requests-oauthlib]\n",
      "   ----------------------- ---------------- 22/38 [posthog]\n",
      "   ----------------------- ---------------- 22/38 [posthog]\n",
      "   ----------------------- ---------------- 22/38 [posthog]\n",
      "   ----------------------- ---------------- 22/38 [posthog]\n",
      "   ---------------- ---------- 23/38 [opentelemetry-exporter-otlp-proto-common]\n",
      "   ------------------------- -------------- 24/38 [markdown-it-py]\n",
      "   ------------------------- -------------- 24/38 [markdown-it-py]\n",
      "   ------------------------- -------------- 24/38 [markdown-it-py]\n",
      "   ------------------------- -------------- 24/38 [markdown-it-py]\n",
      "   ------------------------- -------------- 24/38 [markdown-it-py]\n",
      "   -------------------------- ------------- 25/38 [importlib-metadata]\n",
      "   --------------------------- ------------ 26/38 [humanfriendly]\n",
      "   --------------------------- ------------ 26/38 [humanfriendly]\n",
      "   ---------------------------- ----------- 27/38 [build]\n",
      "   ----------------------------- ---------- 28/38 [rich]\n",
      "   ----------------------------- ---------- 28/38 [rich]\n",
      "   ----------------------------- ---------- 28/38 [rich]\n",
      "   ----------------------------- ---------- 28/38 [rich]\n",
      "   ----------------------------- ---------- 28/38 [rich]\n",
      "   ----------------------------- ---------- 28/38 [rich]\n",
      "   ----------------------------- ---------- 28/38 [rich]\n",
      "   ------------------------------ --------- 29/38 [opentelemetry-api]\n",
      "   ------------------------------ --------- 29/38 [opentelemetry-api]\n",
      "   ------------------------------ --------- 29/38 [opentelemetry-api]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   ------------------------------- -------- 30/38 [kubernetes]\n",
      "   -------------------------------- ------- 31/38 [coloredlogs]\n",
      "   --------------------------------- ------ 32/38 [typer]\n",
      "   --------------------------------- ------ 32/38 [typer]\n",
      "   ---------------------------- ---- 33/38 [opentelemetry-semantic-conventions]\n",
      "   ---------------------------- ---- 33/38 [opentelemetry-semantic-conventions]\n",
      "   ---------------------------- ---- 33/38 [opentelemetry-semantic-conventions]\n",
      "   ---------------------------- ---- 33/38 [opentelemetry-semantic-conventions]\n",
      "   ---------------------------- ---- 33/38 [opentelemetry-semantic-conventions]\n",
      "   ---------------------------- ---- 33/38 [opentelemetry-semantic-conventions]\n",
      "   ---------------------------- ---- 33/38 [opentelemetry-semantic-conventions]\n",
      "   ---------------------------- ---- 33/38 [opentelemetry-semantic-conventions]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ----------------------------------- ---- 34/38 [onnxruntime]\n",
      "   ------------------------------------ --- 35/38 [opentelemetry-sdk]\n",
      "   ------------------------------------ --- 35/38 [opentelemetry-sdk]\n",
      "   ------------------------------------ --- 35/38 [opentelemetry-sdk]\n",
      "   ------------------------------------ --- 35/38 [opentelemetry-sdk]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   -------------------------------------- - 37/38 [chromadb]\n",
      "   ---------------------------------------- 38/38 [chromadb]\n",
      "\n",
      "Successfully installed backoff-2.2.1 bcrypt-4.3.0 build-1.3.0 chromadb-1.0.17 coloredlogs-15.0.1 durationpy-0.10 flatbuffers-25.2.10 httptools-0.6.4 humanfriendly-10.0 importlib-metadata-8.7.0 importlib-resources-6.5.2 kubernetes-33.1.0 markdown-it-py-4.0.0 mdurl-0.1.2 mmh3-5.2.0 oauthlib-3.3.1 onnxruntime-1.22.1 opentelemetry-api-1.36.0 opentelemetry-exporter-otlp-proto-common-1.36.0 opentelemetry-exporter-otlp-proto-grpc-1.36.0 opentelemetry-proto-1.36.0 opentelemetry-sdk-1.36.0 opentelemetry-semantic-conventions-0.57b0 overrides-7.7.0 posthog-5.4.0 pybase64-1.4.2 pypika-0.48.9 pyproject_hooks-1.2.0 pyreadline3-3.5.4 requests-oauthlib-2.0.0 rich-14.1.0 shellingham-1.5.4 typer-0.16.0 uvicorn-0.35.0 watchfiles-1.1.0 websocket-client-1.8.0 websockets-15.0.1 zipp-3.23.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[notice] A new release of pip is available: 25.1.1 -> 25.2\n",
      "[notice] To update, run: python.exe -m pip install --upgrade pip\n"
     ]
    }
   ],
   "source": [
    "%pip install pypdf\n",
    "%pip install pdfplumber\n",
    "%pip install chromadb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60435a4a",
   "metadata": {},
   "source": [
    "#### 1. Document Loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ee87418c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Using PDFPlumber lebo101.pdf: Limit reached while decompressing. 20776104 bytes remaining.\n",
      " Using PDFPlumber lebo102.pdf: Limit reached while decompressing. 13198646 bytes remaining.\n",
      "Total documents loaded: 1940\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from langchain_community.document_loaders import PyPDFLoader, PDFPlumberLoader\n",
    "\n",
    "folder_path = \"./\"\n",
    "\n",
    "docs = []\n",
    "for file in os.listdir(folder_path):\n",
    "    if file.endswith(\".pdf\"):\n",
    "        try:\n",
    "            loader = PyPDFLoader(os.path.join(folder_path, file))\n",
    "            docs.extend(loader.load())\n",
    "        except Exception as e:\n",
    "            loader = PDFPlumberLoader('./lebo101.pdf')\n",
    "            docs.extend(loader.load())\n",
    "            print(f\" Using PDFPlumber {file}: {e}\")\n",
    "\n",
    "print(f\"Total documents loaded: {len(docs)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "55e826fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "file_path = \"data.pkl\"\n",
    "\n",
    "with open(file_path, 'wb') as file:  # 'wb' for write binary\n",
    "    pickle.dump(docs, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ffe9885e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Document(metadata={'producer': 'PDF Printer / www.bullzip.com / FG / Freeware Edition (max 10 users)', 'creator': 'Bullzip PDF Printer (12.2.0.2905)', 'creationdate': '2022-10-17T11:48:38-07:00', 'author': '205DTP3', 'moddate': '2025-04-08T10:06:44+05:30', 'title': 'D:\\\\TEXTBOOKS\\\\RATIONALISED 20222-23\\\\Neha\\\\12086  Chemistry Part_II\\\\1 Source Files\\\\12086  Chemistry Part_II\\\\Unit_9\\\\unit_9.pmd', 'source': './lech204.pdf', 'total_pages': 22, 'page': 12, 'page_label': '13'}, page_content='271 Amines\\nS\\nO\\nO\\nCl + H N C H2 5 S\\nO\\nO\\nN C H2 5 + HCl\\nN,N-Diethylbenzenesulphonamide\\nC H2 5 C H2 5\\n4. Carbylamine reaction\\nAliphatic and aromatic primary amines on heating with chloroform\\nand ethanolic potassium hydroxide form isocyanides or carbylamines\\nwhich are foul smelling substances.  Secondary and tertiary amines\\ndo not show this reaction.  This reaction is known as carbylamine\\nreaction or isocyanide test and is used as a test for primary amines.\\n5. Reaction with nitrous acid\\nThree classes of amines react differently with nitrous acid which is\\nprepared in situ from a mineral acid and sodium nitrite.\\n(a) Primary aliphatic amines react with nitrous acid to form aliphatic\\ndiazonium salts which being unstable, liberate nitrogen gas\\nquantitatively and alcohols. Quantitative evolution of nitrogen is\\nused in estimation of amino acids and proteins.\\n(b) Aromatic amines react with nitrous acid at low temperatures\\n(273-278 K) to form diazonium salts, a very important class of\\ncompounds used for synthesis of a variety of aromatic compounds\\ndiscussed in Section 9.7.\\nSecondary and tertiary amines react with nitrous acid in a\\ndifferent manner.\\n6. Reaction with arylsulphonyl chloride\\nBenzenesulphonyl chloride (C 6H5SO2Cl), which is also known as\\nHinsbergs reagent, reacts with primary and secondary amines to\\nform sulphonamides.\\n(a) The reaction of benzenesulphonyl chloride with primary amine\\nyields N-ethylbenzenesulphonyl amide.\\nThe hydrogen attached to nitrogen in sulphonamide is strongly\\nacidic due to the presence of strong electron withdrawing sulphonyl\\ngroup. Hence, it is soluble in alkali.\\n(b) In the reaction with secondary amine, N,N-diethyl-\\nbenzenesulphonamide is formed.\\nReprint 2025-26')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "docs[1500]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9971c370",
   "metadata": {},
   "source": [
    "#### 2. Vector Store"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "726ad1b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_openai import OpenAIEmbeddings, ChatOpenAI\n",
    "from langchain.vectorstores import Chroma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c789e279",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\LoG1c\\AppData\\Local\\Temp\\ipykernel_7596\\2641773464.py:1: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.\n",
      "  vector_store = Chroma(\n"
     ]
    }
   ],
   "source": [
    "vector_store = Chroma(\n",
    "    embedding_function=OpenAIEmbeddings(),\n",
    "    persist_directory='my_books_db',\n",
    "    collection_name='sample'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "bb2f076c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add documents\n",
    "vector_store.add_documents(docs[:400])\n",
    "vector_store.persist()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a4cdcb6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add documents\n",
    "vector_store.add_documents(docs[400:700])\n",
    "vector_store.persist()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "d2fb8422",
   "metadata": {},
   "outputs": [],
   "source": [
    "vector_store.add_documents(docs[700:1000])\n",
    "vector_store.persist()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "42aad379",
   "metadata": {},
   "outputs": [],
   "source": [
    "vector_store.add_documents(docs[1000:1300])\n",
    "vector_store.persist()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "dd1f4597",
   "metadata": {},
   "outputs": [],
   "source": [
    "vector_store.add_documents(docs[1300:1600])\n",
    "vector_store.persist()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "cdba5bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "vector_store.add_documents(docs[1600:])\n",
    "vector_store.persist()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ab064dbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total documents loaded: 1940\n"
     ]
    }
   ],
   "source": [
    "all_docs = vector_store.get()\n",
    "print(\"Total documents loaded:\", len(all_docs[\"ids\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7e660d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "retriever = vector_store.as_retriever(search_type=\"similarity\", search_kwargs={\"k\": 4})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3a0934b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(metadata={'total_pages': 14, 'creationdate': '2022-10-14T14:18:24+00:00', 'page_label': '6', 'moddate': '2025-04-30T15:23:29+05:30', 'page': 5, 'source': './kebo103.pdf', 'title': 'ch-3.pmd', 'creator': 'PageMaker 7.0', 'producer': 'GPL Ghostscript 8.15', 'author': 'Administrator'}, page_content='28 BIOLOGY\\ngametes. Sexual reproduction is oogamous and accompanied by complex\\npost fertilisation developments. The common members are: Polysiphonia,\\nPorphyra (Figure 3.1c), Gracilaria and Gelidium.\\n3.2 BRYOPHYTES\\nBryophytes include the various mosses and liverworts that are found\\ncommonly growing in moist shaded areas in the hills (Figure 3.2).\\nArchegoniophore\\n(a) (b)\\n(c)\\n(d)\\nAntheridiophore\\nCapsule\\nAntheridial\\nbranch Branches\\nArchegonial\\nbranch\\nSeta\\nSporophyte\\nGametophyte\\nLeaves\\nMain axis\\nRhizoids\\nGemma cup\\nRhizoids\\nGemma cup\\nRhizoids\\nFigure 3.2 Bryophytes: A liverwort  Marchantia (a) Female thallus  (b) Male thallus\\nMosses   (c) Funaria, gametophyte and sporophyte  (d) Sphagnum\\ngametophyte\\nReprint 2025-26'),\n",
       " Document(metadata={'page': 11, 'total_pages': 25, 'file_path': './lebo101.pdf', 'source': './lebo101.pdf', 'CreationDate': \"D:20250430152404+05'30'\", 'Producer': 'iLovePDF', 'ModDate': \"D:20250430152404+05'30'\"}, page_content='BIOLOGY\\nlie close to each other so that self-pollination\\ncan occur. Some plants such as Viola\\n(common pansy), Oxalis, and Commelina\\nproduce two types of flowers \\nchasmogamous flowers which are similar to\\nflowers of other species with exposed anthers\\nand stigma, and cleistogamous flowers which\\ndo not open at all (Figure 1.9c). In such flowers,\\nthe anthers and stigma lie close to each other.\\n(a) When anthers dehisce in the flower buds,\\npollen grains come in contact with the stigma\\nto effect pollination. Thus, cleistogamous\\nflowers are invariably autogamous as there is\\nno chance of cross-pollen landing on the\\nstigma. Cleistogamous flowers produce\\nassured seed-set even in the absence of\\npollinators. Do you think that cleistogamy is\\nadvantageous or disadvantageous to the\\nplant? Why?\\n(b) (ii) GGGGGeeeeeiiiiitttttooooonnnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nthe anther to the stigma of another flower of\\nthe same plant. Although geitonogamy is\\nfunctionally cross-pollination involving a\\npollinating agent, genetically it is similar to\\nautogamy since the pollen grains come from\\nthe same plant.\\n(iii) XXXXXeeeeennnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nanther to the stigma of a different plant (Figure\\n1.9b). This is the only type of pollination which\\nduring pollination brings genetically different\\ntypes of pollen grains to the stigma.\\nAgents of Pollination : Plants use two abiotic (wind\\nand water) and one biotic (animals) agents to achieve\\npollination. Majority of plants use biotic agents for\\npollination. Only a small proportion of plants use\\n12\\n(c) abiotic agents. Pollen grains coming in contact with\\nthe stigma is a chance factor in both wind and water\\npollination. To compensate for this uncertainties and\\nassociated loss of pollen grains, the flowers produce\\nFigure 1.9 (a) Self-pollinated flowers;\\n(b)Cross pollinated flowers; enormous amount of pollen when compared to the\\n(c)Cleistogamous flowers\\nnumber of ovules available for pollination.\\nReprint 2025-26\\n'),\n",
       " Document(metadata={'ModDate': \"D:20250430152404+05'30'\", 'total_pages': 25, 'source': './lebo101.pdf', 'CreationDate': \"D:20250430152404+05'30'\", 'page': 11, 'Producer': 'iLovePDF', 'file_path': './lebo101.pdf'}, page_content='BIOLOGY\\nlie close to each other so that self-pollination\\ncan occur. Some plants such as Viola\\n(common pansy), Oxalis, and Commelina\\nproduce two types of flowers \\nchasmogamous flowers which are similar to\\nflowers of other species with exposed anthers\\nand stigma, and cleistogamous flowers which\\ndo not open at all (Figure 1.9c). In such flowers,\\nthe anthers and stigma lie close to each other.\\n(a) When anthers dehisce in the flower buds,\\npollen grains come in contact with the stigma\\nto effect pollination. Thus, cleistogamous\\nflowers are invariably autogamous as there is\\nno chance of cross-pollen landing on the\\nstigma. Cleistogamous flowers produce\\nassured seed-set even in the absence of\\npollinators. Do you think that cleistogamy is\\nadvantageous or disadvantageous to the\\nplant? Why?\\n(b) (ii) GGGGGeeeeeiiiiitttttooooonnnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nthe anther to the stigma of another flower of\\nthe same plant. Although geitonogamy is\\nfunctionally cross-pollination involving a\\npollinating agent, genetically it is similar to\\nautogamy since the pollen grains come from\\nthe same plant.\\n(iii) XXXXXeeeeennnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nanther to the stigma of a different plant (Figure\\n1.9b). This is the only type of pollination which\\nduring pollination brings genetically different\\ntypes of pollen grains to the stigma.\\nAgents of Pollination : Plants use two abiotic (wind\\nand water) and one biotic (animals) agents to achieve\\npollination. Majority of plants use biotic agents for\\npollination. Only a small proportion of plants use\\n12\\n(c) abiotic agents. Pollen grains coming in contact with\\nthe stigma is a chance factor in both wind and water\\npollination. To compensate for this uncertainties and\\nassociated loss of pollen grains, the flowers produce\\nFigure 1.9 (a) Self-pollinated flowers;\\n(b)Cross pollinated flowers; enormous amount of pollen when compared to the\\n(c)Cleistogamous flowers\\nnumber of ovules available for pollination.\\nReprint 2025-26\\n'),\n",
       " Document(metadata={'ModDate': \"D:20250430152404+05'30'\", 'file_path': './lebo101.pdf', 'source': './lebo101.pdf', 'total_pages': 25, 'CreationDate': \"D:20250430152404+05'30'\", 'page': 22, 'Producer': 'iLovePDF'}, page_content='HSEX\\nU\\nU\\nM\\nA\\nA\\nL\\nN\\nR REP\\nE\\nR\\nP\\nO\\nR\\nD\\nO\\nU\\nD\\nC\\nU\\nT\\nC\\nIO\\nT\\nN\\nIO\\nI\\nN\\nN FLOWERING PLANTS\\nvarieties some of the nucellar cells surrounding the embryo sac start\\ndividing, protrude into the embryo sac and develop into the embryos. In\\nsuch species each ovule contains many embryos. Occurrence of more\\nthan one embryo in a seed is referred to as polyembryony. Take out\\nsome seeds of orange and squeeze them. Observe the many embryos of\\ndifferent sizes and shapes from each seed. Count the number of embryos\\nin each seed. What would be the genetic nature of apomictic embryos?\\nCan they be called clones?\\nHybrid varieties of several of our food and vegetable crops are being\\nextensively cultivated. Cultivation of hybrids has tremendously increased\\nproductivity. One of the problems of hybrids is that hybrid seeds have\\nto be produced every year. If the seeds collected from hybrids are sown,\\nthe plants in the progeny will segregate and do not maintain hybrid\\ncharacters. Production of hybrid seeds is costly and hence the cost of\\nhybrid seeds become too expensive for the farmers. If these hybrids are\\nmade into apomicts, there is no segregation of characters in the hybrid\\nprogeny. Then the farmers can keep on using the hybrid seeds to raise\\nnew crop year after year and he does not have to buy hybrid seeds every\\nyear. Because of the importance of apomixis in hybrid seed industry,\\nactive research is going on in many laboratories around the world to\\nunderstand the genetics of apomixis and to transfer apomictic genes\\ninto hybrid varieties.\\nSUMMARY\\nFlowers are the seat of sexual reproduction in angiosperms. In the flower,\\nandroecium consisting of stamens represents the male reproductive\\norgans and gynoecium consisting of pistils represents the female\\nreproductive organs.\\nA typical anther is bilobed, dithecous and tetrasporangiate. Pollen\\ngrains develop inside the microsporangia. Four wall layers, the\\nepidermis, endothecium, middle layers and the tapetum surround the\\nmicrosporangium. Cells of the sporogenous tissue lying in the centre of\\nthe microsporangium, undergo meiosis (microsporogenesis) to form\\ntetrads of microspores. Individual microspores mature into pollen grains.\\nPollen grains represents the male gametophytic generation. The\\npollen grains have a two-layered wall, the outer exine and inner intine.\\nThe exine is made up of sporopollenin and has germ pores. Pollen grains\\nmay have two cells (a vegetative cell and generative cell) or three cells (a\\nvegetative cell and two male gametes) at the time of shedding.\\n23\\nThe pistil has three parts  the stigma, style and the ovary. Ovules\\nare present in the ovary. The ovules have a stalk called funicle, protective\\nintegument(s), and an opening called micropyle. The central tissue is\\nthe nucellus in which the archesporium differentiates. A cell of the\\narchesporium, the megaspore mother cell divides meiotically and one of\\nthe megaspores forms the embryo sac (the female gametophyte). The\\nmature embryo sac is 7-celled and 8-nucleate. At the micropylar end is\\nReprint 2025-26\\n')]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "retriever.invoke('examples of isogamous reproduction?')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fd291e01",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.retrievers.multi_query import MultiQueryRetriever\n",
    "\n",
    "multiquery_retriever = MultiQueryRetriever.from_llm(\n",
    "    retriever=vector_store.as_retriever(search_kwargs={\"k\": 5}),\n",
    "    llm=ChatOpenAI(model=\"gpt-4o-mini\", temperature=0)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8ea38e46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(metadata={'page_label': '6', 'producer': 'GPL Ghostscript 8.15', 'creator': 'PageMaker 7.0', 'source': './kebo103.pdf', 'author': 'Administrator', 'creationdate': '2022-10-14T14:18:24+00:00', 'page': 5, 'title': 'ch-3.pmd', 'total_pages': 14, 'moddate': '2025-04-30T15:23:29+05:30'}, page_content='28 BIOLOGY\\ngametes. Sexual reproduction is oogamous and accompanied by complex\\npost fertilisation developments. The common members are: Polysiphonia,\\nPorphyra (Figure 3.1c), Gracilaria and Gelidium.\\n3.2 BRYOPHYTES\\nBryophytes include the various mosses and liverworts that are found\\ncommonly growing in moist shaded areas in the hills (Figure 3.2).\\nArchegoniophore\\n(a) (b)\\n(c)\\n(d)\\nAntheridiophore\\nCapsule\\nAntheridial\\nbranch Branches\\nArchegonial\\nbranch\\nSeta\\nSporophyte\\nGametophyte\\nLeaves\\nMain axis\\nRhizoids\\nGemma cup\\nRhizoids\\nGemma cup\\nRhizoids\\nFigure 3.2 Bryophytes: A liverwort  Marchantia (a) Female thallus  (b) Male thallus\\nMosses   (c) Funaria, gametophyte and sporophyte  (d) Sphagnum\\ngametophyte\\nReprint 2025-26'),\n",
       " Document(metadata={'source': './lebo101.pdf', 'file_path': './lebo101.pdf', 'Producer': 'iLovePDF', 'total_pages': 25, 'ModDate': \"D:20250430152404+05'30'\", 'page': 0, 'CreationDate': \"D:20250430152404+05'30'\"}, page_content='Chapter 1 Biology in essence is the story of life on earth. While individual\\nSexual Reproduction in organisms die without fail, species continue to live through\\nflowering Plants millions of years unless threatened by natural or anthropogenic\\nextinction. Reproduction becomes a vital process without\\nChapter 2 which species cannot survive for long. Each individual leaves\\nHuman Reproduction its progeny by asexual or sexual means. Sexual mode of\\nreproduction enables creation of new variants, so that survival\\nChapter 3\\nadvantage is enhanced. This unit explains the details of\\nReproductive Health\\nreproductive processes in flowering plants and humans as easy\\nto relate representative examples. A related perspective on\\nhuman reproductive health and how reproductive ill health\\ncan be avoided is also presented to complete our\\nunderstanding of biology of reproduction.\\nReprint 2025-26\\n'),\n",
       " Document(metadata={'page_label': '8', 'title': 'ch-3.pmd', 'total_pages': 14, 'producer': 'GPL Ghostscript 8.15', 'creationdate': '2022-10-14T14:18:24+00:00', 'author': 'Administrator', 'source': './kebo103.pdf', 'moddate': '2025-04-30T15:23:29+05:30', 'page': 7, 'creator': 'PageMaker 7.0'}, page_content='30 BIOLOGY\\norgans are produced either on the same or on different thalli. The\\nsporophyte is differentiated into a foot, seta and capsule. After meiosis,\\nspores are produced within the capsule. These spores germinate to form\\nfree-living gametophytes.\\n3.2.2 Mosses\\nThe predominant stage of the life cycle of a moss is the gametophyte which\\nconsists of two stages. The first stage is the protonema stage, which\\ndevelops directly from a spore. It is a creeping, green, branched and\\nfrequently filamentous stage. The second stage is the leafy stage, which\\ndevelops from the secondary protonema as a lateral bud. They consist of\\nupright, slender axes bearing spirally arranged leaves. They are attached\\nto the soil through multicellular and branched rhizoids. This stage bears\\nthe sex organs.\\nVegetative reproduction in mosses is by fragmentation and budding\\nin the secondary protonema. In sexual reproduction, the sex organs\\nantheridia and archegonia are produced at the apex of the leafy shoots.\\nAfter fertilisation, the zygote develops into a sporophyte, consisting of a\\nfoot, seta and capsule. The sporophyte in mosses is more elaborate than\\nthat in liverworts. The capsule contains spores. Spores are formed after\\nmeiosis. The mosses have an elaborate mechanism of spore dispersal.\\nCommon examples of mosses are \\nFunaria, Polytrichum and Sphagnum\\n(Figure 3.2).\\n3.3 PTERIDOPHYTES\\nThe Pteridophytes include horsetails and ferns. Pteridophytes are used\\nfor medicinal purposes and as soil-binders. They are also frequently grown\\nas ornamentals. Evolutionarily, they are the first terrestrial plants to\\npossess vascular tissues  xylem and phloem. You shall study more about\\nthese tissues in Chapter 6. The pteridophytes are found in cool, damp,\\nshady places though some may flourish well in sandy-soil conditions.\\nYou may recall that in bryophytes the dominant phase in the life\\ncycle is the gametophytic plant body. However , in pteridophytes, the\\nmain plant body is a sporophyte which is differentiated into true root,\\nstem and leaves (Figure 3.3). These organs possess well-differentiated\\nvascular tissues. The leaves in pteridophyta are small (microphylls) as\\nin Selaginella or large (macrophylls) as in ferns. The sporophytes bear\\nsporangia that are subtended by leaf-like appendages called\\nsporophylls . In some cases sporophylls may form distinct compact\\nstructures called strobili or cones ( Selaginella, Equisetum ). The\\nsporangia produce spores by meiosis in spore mother cells. The spores\\ngerminate to give rise to inconspicuous, small but multicellular ,\\nReprint 2025-26'),\n",
       " Document(metadata={'Producer': 'iLovePDF', 'total_pages': 25, 'file_path': './lebo101.pdf', 'source': './lebo101.pdf', 'page': 24, 'CreationDate': \"D:20250430152404+05'30'\", 'ModDate': \"D:20250430152404+05'30'\"}, page_content='HSEX\\nU\\nU\\nM\\nA\\nA\\nL\\nN\\nR REP\\nE\\nR\\nP\\nO\\nR\\nD\\nO\\nU\\nD\\nC\\nU\\nT\\nC\\nIO\\nT\\nN\\nIO\\nI\\nN\\nN FLOWERING PLANTS\\n7. What are chasmogamous flowers? Can cross-pollination occur in\\ncleistogamous flowers? Give reasons for your answer.\\n8. Mention two strategies evolved to prevent self-pollination in flowers.\\n9. What is self-incompatibility? Why does self-pollination not lead to seed\\nformation in self-incompatible species?\\n10. What is bagging technique? How is it useful in a plant breeding\\nprogramme?\\n11. What is triple fusion? Where and how does it take place? Name the\\nnuclei involved in triple fusion.\\n12. Why do you think the zygote is dormant for sometime in a fertilised\\novule?\\n13. Differentiate between:\\n(a) hypocotyl and epicotyl;\\n(b) coleoptile and coleorrhiza;\\n(c) integument and testa;\\n(d) perisperm and pericarp.\\n14. Why is apple called a false fruit? Which part(s) of the flower forms the\\nfruit?\\n15. What is meant by emasculation? When and why does a plant breeder\\nemploy this technique?\\n16. If one can induce parthenocarpy through the application of growth\\nsubstances, which fruits would you select to induce parthenocarpy\\nand why?\\n17. Explain the role of tapetum in the formation of pollen-grain wall.\\n18. What is apomixis and what is its importance?\\n25\\nReprint 2025-26\\n'),\n",
       " Document(metadata={'page': 11, 'Producer': 'iLovePDF', 'file_path': './lebo101.pdf', 'ModDate': \"D:20250430152404+05'30'\", 'total_pages': 25, 'CreationDate': \"D:20250430152404+05'30'\", 'source': './lebo101.pdf'}, page_content='BIOLOGY\\nlie close to each other so that self-pollination\\ncan occur. Some plants such as Viola\\n(common pansy), Oxalis, and Commelina\\nproduce two types of flowers \\nchasmogamous flowers which are similar to\\nflowers of other species with exposed anthers\\nand stigma, and cleistogamous flowers which\\ndo not open at all (Figure 1.9c). In such flowers,\\nthe anthers and stigma lie close to each other.\\n(a) When anthers dehisce in the flower buds,\\npollen grains come in contact with the stigma\\nto effect pollination. Thus, cleistogamous\\nflowers are invariably autogamous as there is\\nno chance of cross-pollen landing on the\\nstigma. Cleistogamous flowers produce\\nassured seed-set even in the absence of\\npollinators. Do you think that cleistogamy is\\nadvantageous or disadvantageous to the\\nplant? Why?\\n(b) (ii) GGGGGeeeeeiiiiitttttooooonnnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nthe anther to the stigma of another flower of\\nthe same plant. Although geitonogamy is\\nfunctionally cross-pollination involving a\\npollinating agent, genetically it is similar to\\nautogamy since the pollen grains come from\\nthe same plant.\\n(iii) XXXXXeeeeennnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nanther to the stigma of a different plant (Figure\\n1.9b). This is the only type of pollination which\\nduring pollination brings genetically different\\ntypes of pollen grains to the stigma.\\nAgents of Pollination : Plants use two abiotic (wind\\nand water) and one biotic (animals) agents to achieve\\npollination. Majority of plants use biotic agents for\\npollination. Only a small proportion of plants use\\n12\\n(c) abiotic agents. Pollen grains coming in contact with\\nthe stigma is a chance factor in both wind and water\\npollination. To compensate for this uncertainties and\\nassociated loss of pollen grains, the flowers produce\\nFigure 1.9 (a) Self-pollinated flowers;\\n(b)Cross pollinated flowers; enormous amount of pollen when compared to the\\n(c)Cleistogamous flowers\\nnumber of ovules available for pollination.\\nReprint 2025-26\\n'),\n",
       " Document(metadata={'creator': 'PageMaker 7.0', 'title': 'ch-3.pmd', 'total_pages': 14, 'creationdate': '2022-10-14T14:18:24+00:00', 'moddate': '2025-04-30T15:23:29+05:30', 'author': 'Administrator', 'source': './kebo103.pdf', 'page_label': '14', 'page': 13, 'producer': 'GPL Ghostscript 8.15'}, page_content='36 BIOLOGY\\n5. Write a note on economic importance of algae and gymnosperms.\\n6. Both gymnosperms and angiosperms bear seeds, then why are they\\nclassified separately?\\n7. What is heterospory? Briefly comment on its significance. Give two\\nexamples.\\n8. Explain briefly the following terms with suitable examples:-\\n(i) protonema\\n(ii) antheridium\\n(iii) archegonium\\n(iv) diplontic\\n(v) sporophyll\\n(vi) isogamy\\n9. Differentiate between the following:-\\n(i) red algae and brown algae\\n(ii) liverworts and moss\\n(iii) homosporous and heterosporous pteridophyte\\n10. Match the following (column I with column II)\\nColumn I Column II\\n(a) Chlamydomonas (i) Moss\\n(b) Cycas (ii) Pteridophyte\\n(c) Selaginella (iii) Algae\\n(d)  Sphagnum (iv) Gymnosperm\\n11. Describe the important characteristics of gymnosperms.\\nReprint 2025-26'),\n",
       " Document(metadata={'ModDate': \"D:20250430152404+05'30'\", 'total_pages': 25, 'source': './lebo101.pdf', 'page': 20, 'file_path': './lebo101.pdf', 'Producer': 'iLovePDF', 'CreationDate': \"D:20250430152404+05'30'\"}, page_content='HSEX\\nU\\nU\\nM\\nA\\nA\\nL\\nN\\nR REP\\nE\\nR\\nP\\nO\\nR\\nD\\nO\\nU\\nD\\nC\\nU\\nT\\nC\\nIO\\nT\\nN\\nIO\\nI\\nN\\nN FLOWERING PLANTS\\n(a)\\n(b)\\nFigure 1.15 (a) Structure of some seeds. (b) False fruits of apple and strawberry\\nin which fruits develop without fertilisation. Such fruits are called\\nparthenocarpic fruits. Banana is one such example. Parthenocarpy can 21\\nbe induced through the application of growth hormones and such fruits\\nare seedless.\\nSeeds offer several advantages to angiosperms. Firstly, since\\nreproductive processes such as pollination and fertilisation are\\nindependent of water, seed formation is more dependable. Also seeds have\\nbetter adaptive strategies for dispersal to new habitats and help the species\\nReprint 2025-26\\n')]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiquery_retriever.invoke('Examples of isogamous reproduction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c9b371ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting rank_bm25\n",
      "  Downloading rank_bm25-0.2.2-py3-none-any.whl.metadata (3.2 kB)\n",
      "Requirement already satisfied: numpy in a:\\work\\langchain\\lang\\lib\\site-packages (from rank_bm25) (2.2.5)\n",
      "Downloading rank_bm25-0.2.2-py3-none-any.whl (8.6 kB)\n",
      "Installing collected packages: rank_bm25\n",
      "Successfully installed rank_bm25-0.2.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install rank_bm25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6c9b76ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(metadata={'ModDate': \"D:20250430152404+05'30'\", 'page': 22, 'source': './lebo101.pdf', 'file_path': './lebo101.pdf', 'CreationDate': \"D:20250430152404+05'30'\", 'Producer': 'iLovePDF', 'total_pages': 25}, page_content='HSEX\\nU\\nU\\nM\\nA\\nA\\nL\\nN\\nR REP\\nE\\nR\\nP\\nO\\nR\\nD\\nO\\nU\\nD\\nC\\nU\\nT\\nC\\nIO\\nT\\nN\\nIO\\nI\\nN\\nN FLOWERING PLANTS\\nvarieties some of the nucellar cells surrounding the embryo sac start\\ndividing, protrude into the embryo sac and develop into the embryos. In\\nsuch species each ovule contains many embryos. Occurrence of more\\nthan one embryo in a seed is referred to as polyembryony. Take out\\nsome seeds of orange and squeeze them. Observe the many embryos of\\ndifferent sizes and shapes from each seed. Count the number of embryos\\nin each seed. What would be the genetic nature of apomictic embryos?\\nCan they be called clones?\\nHybrid varieties of several of our food and vegetable crops are being\\nextensively cultivated. Cultivation of hybrids has tremendously increased\\nproductivity. One of the problems of hybrids is that hybrid seeds have\\nto be produced every year. If the seeds collected from hybrids are sown,\\nthe plants in the progeny will segregate and do not maintain hybrid\\ncharacters. Production of hybrid seeds is costly and hence the cost of\\nhybrid seeds become too expensive for the farmers. If these hybrids are\\nmade into apomicts, there is no segregation of characters in the hybrid\\nprogeny. Then the farmers can keep on using the hybrid seeds to raise\\nnew crop year after year and he does not have to buy hybrid seeds every\\nyear. Because of the importance of apomixis in hybrid seed industry,\\nactive research is going on in many laboratories around the world to\\nunderstand the genetics of apomixis and to transfer apomictic genes\\ninto hybrid varieties.\\nSUMMARY\\nFlowers are the seat of sexual reproduction in angiosperms. In the flower,\\nandroecium consisting of stamens represents the male reproductive\\norgans and gynoecium consisting of pistils represents the female\\nreproductive organs.\\nA typical anther is bilobed, dithecous and tetrasporangiate. Pollen\\ngrains develop inside the microsporangia. Four wall layers, the\\nepidermis, endothecium, middle layers and the tapetum surround the\\nmicrosporangium. Cells of the sporogenous tissue lying in the centre of\\nthe microsporangium, undergo meiosis (microsporogenesis) to form\\ntetrads of microspores. Individual microspores mature into pollen grains.\\nPollen grains represents the male gametophytic generation. The\\npollen grains have a two-layered wall, the outer exine and inner intine.\\nThe exine is made up of sporopollenin and has germ pores. Pollen grains\\nmay have two cells (a vegetative cell and generative cell) or three cells (a\\nvegetative cell and two male gametes) at the time of shedding.\\n23\\nThe pistil has three parts  the stigma, style and the ovary. Ovules\\nare present in the ovary. The ovules have a stalk called funicle, protective\\nintegument(s), and an opening called micropyle. The central tissue is\\nthe nucellus in which the archesporium differentiates. A cell of the\\narchesporium, the megaspore mother cell divides meiotically and one of\\nthe megaspores forms the embryo sac (the female gametophyte). The\\nmature embryo sac is 7-celled and 8-nucleate. At the micropylar end is\\nReprint 2025-26\\n'),\n",
       " Document(metadata={'Producer': 'iLovePDF', 'page': 11, 'total_pages': 25, 'source': './lebo101.pdf', 'CreationDate': \"D:20250430152404+05'30'\", 'ModDate': \"D:20250430152404+05'30'\", 'file_path': './lebo101.pdf'}, page_content='BIOLOGY\\nlie close to each other so that self-pollination\\ncan occur. Some plants such as Viola\\n(common pansy), Oxalis, and Commelina\\nproduce two types of flowers \\nchasmogamous flowers which are similar to\\nflowers of other species with exposed anthers\\nand stigma, and cleistogamous flowers which\\ndo not open at all (Figure 1.9c). In such flowers,\\nthe anthers and stigma lie close to each other.\\n(a) When anthers dehisce in the flower buds,\\npollen grains come in contact with the stigma\\nto effect pollination. Thus, cleistogamous\\nflowers are invariably autogamous as there is\\nno chance of cross-pollen landing on the\\nstigma. Cleistogamous flowers produce\\nassured seed-set even in the absence of\\npollinators. Do you think that cleistogamy is\\nadvantageous or disadvantageous to the\\nplant? Why?\\n(b) (ii) GGGGGeeeeeiiiiitttttooooonnnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nthe anther to the stigma of another flower of\\nthe same plant. Although geitonogamy is\\nfunctionally cross-pollination involving a\\npollinating agent, genetically it is similar to\\nautogamy since the pollen grains come from\\nthe same plant.\\n(iii) XXXXXeeeeennnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nanther to the stigma of a different plant (Figure\\n1.9b). This is the only type of pollination which\\nduring pollination brings genetically different\\ntypes of pollen grains to the stigma.\\nAgents of Pollination : Plants use two abiotic (wind\\nand water) and one biotic (animals) agents to achieve\\npollination. Majority of plants use biotic agents for\\npollination. Only a small proportion of plants use\\n12\\n(c) abiotic agents. Pollen grains coming in contact with\\nthe stigma is a chance factor in both wind and water\\npollination. To compensate for this uncertainties and\\nassociated loss of pollen grains, the flowers produce\\nFigure 1.9 (a) Self-pollinated flowers;\\n(b)Cross pollinated flowers; enormous amount of pollen when compared to the\\n(c)Cleistogamous flowers\\nnumber of ovules available for pollination.\\nReprint 2025-26\\n'),\n",
       " Document(metadata={'producer': 'GPL Ghostscript 8.15', 'creator': 'PageMaker 7.0', 'creationdate': '2022-10-14T14:16:33+00:00', 'author': 'Administrator', 'moddate': '2025-04-30T15:23:27+05:30', 'title': 'ch-2.pmd', 'source': './kebo102.pdf', 'total_pages': 13, 'page': 7, 'page_label': '8'}, page_content='BIOLOGICAL CLASSIFICATION 17\\ncalled conidia or sporangiospores or zoospores, and sexual reproduction\\nis by oospores, ascospores and basidiospores. The various spores are\\nproduced in distinct structures called fruiting bodies. The sexual cycle\\ninvolves the following  three steps:\\n(i) Fusion of protoplasms between two motile or non-motile gametes\\ncalled plasmogamy.\\n(ii) Fusion of two nuclei called karyogamy.\\n(iii) Meiosis in zygote resulting in haploid spores.\\nWhen a fungus reproduces sexually, two haploid\\nhyphae of compatible mating types come together and\\nfuse. In some fungi the fusion of two haploid cells\\nimmediately results in diploid cells (2n). However, in other\\nfungi (ascomycetes and basidiomycetes), an intervening\\ndikaryotic stage (n + n, i.e., two nuclei per cell) occurs;\\nsuch a condition is called a dikaryon and the phase is\\ncalled dikaryophase of fungus. Later, the parental nuclei\\nfuse and the cells become diploid.  The fungi form fruiting\\nbodies in which reduction division occurs, leading to\\nformation of haploid spores.\\nThe morphology of the mycelium, mode of spore\\nformation and fruiting bodies  form the basis for the\\ndivision of the kingdom  into various classes.\\n2.3.1 Phycomycetes\\nMembers of phycomycetes are found in aquatic habitats\\nand on decaying wood in moist and damp places or as\\nobligate parasites on plants. The mycelium is aseptate\\nand coenocytic. Asexual reproduction takes place by\\nzoospores (motile) or by aplanospores (non-motile). These\\nspores are endogenously produced in sporangium. A\\nzygospore is formed by fusion of two gametes. These\\ngametes are similar in morphology (isogamous) or\\ndissimilar (anisogamous or oogamous). Some common\\nexamples are Mucor (Figure 2.5a), Rhizopus (the bread\\nmould mentioned earlier) and Albugo (the parasitic fungi\\non mustard).\\n2.3.2 Ascomycetes\\nCommonly known as sac-fungi, the ascomycetes  are mostly\\nmulticellular, e.g., Penicillium, or rarely unicellular, e.g., yeast\\n(Saccharomyces). They are saprophytic, decomposers,\\nparasitic or coprophilous (growing on dung). Mycelium\\nFigure 2.5 Fungi: (a) Mucor\\n(b) Aspergillus (c) Agaricus\\n(c)\\n(a)\\n(b)\\nReprint 2025-26'),\n",
       " Document(metadata={'page': 5, 'moddate': '2025-04-30T15:23:29+05:30', 'page_label': '6', 'source': './kebo103.pdf', 'total_pages': 14, 'title': 'ch-3.pmd', 'creationdate': '2022-10-14T14:18:24+00:00', 'creator': 'PageMaker 7.0', 'producer': 'GPL Ghostscript 8.15', 'author': 'Administrator'}, page_content='28 BIOLOGY\\ngametes. Sexual reproduction is oogamous and accompanied by complex\\npost fertilisation developments. The common members are: Polysiphonia,\\nPorphyra (Figure 3.1c), Gracilaria and Gelidium.\\n3.2 BRYOPHYTES\\nBryophytes include the various mosses and liverworts that are found\\ncommonly growing in moist shaded areas in the hills (Figure 3.2).\\nArchegoniophore\\n(a) (b)\\n(c)\\n(d)\\nAntheridiophore\\nCapsule\\nAntheridial\\nbranch Branches\\nArchegonial\\nbranch\\nSeta\\nSporophyte\\nGametophyte\\nLeaves\\nMain axis\\nRhizoids\\nGemma cup\\nRhizoids\\nGemma cup\\nRhizoids\\nFigure 3.2 Bryophytes: A liverwort  Marchantia (a) Female thallus  (b) Male thallus\\nMosses   (c) Funaria, gametophyte and sporophyte  (d) Sphagnum\\ngametophyte\\nReprint 2025-26'),\n",
       " Document(metadata={'producer': 'GPL Ghostscript 8.15', 'creator': 'PageMaker 7.0', 'creationdate': '2022-10-14T14:16:33+00:00', 'author': 'Administrator', 'moddate': '2025-04-30T15:23:27+05:30', 'title': 'ch-2.pmd', 'source': './kebo102.pdf', 'total_pages': 13, 'page': 8, 'page_label': '9'}, page_content='18 BIOLOGY\\nis branched and septate. The asexual spores are conidia produced\\nexogenously on the special mycelium called conidiophores. Conidia on\\ngermination produce mycelium. Sexual spores are called ascospores\\nwhich are produced endogenously in sac like asci (singular ascus). These\\nasci are arranged in different types of fruiting bodies called ascocarps.\\nSome examples are Aspergillus (Figure 2.5b), Claviceps and Neurospora.\\nNeurospora  is used extensively in biochemical and genetic work. Many\\nmembers like morels and truffles are edible and are considered  delicacies.\\n2.3.3 Basidiomycetes\\nCommonly known forms of basidiomycetes are mushrooms, bracket\\nfungi or puffballs. They grow in soil, on logs and tre e stumps and in\\nliving plant bodies as parasites, e.g., rusts and smuts. The mycelium is\\nbranched and septate. The asexual spores are generally not found, but\\nvegetative reproduction by fragmentation is common. The sex organs\\nare absent, but plasmogamy is brought about by fusion of two vegetative\\nor somatic cells of different strains or genotypes. The resultant structure\\nis dikaryotic which ultimately gives rise to basidium. Karyogamy and\\nmeiosis take place in the basidium producing four basidiospores. The\\nbasidiospores are exogenously produced on the basidium (pl.: basidia).\\nThe basidia are arranged in fruiting bodies called basidiocarps. Some\\ncommon members are Agaricus (mushroom) (Figure 2.5c), Ustilago \\n(smut)\\nand Puccinia (rust fungus).\\n2.3.4 Deuteromycetes\\nCommonly known as imperfect fungi because only the asexual or\\nvegetative phases of these fungi are known.  When the sexual forms of\\nthese fungi were discovered they were moved into classes they rightly\\nbelong to.  It is also possible that the asexual and vegetative stage have\\nbeen given one name (and placed under deuteromycetes) and the sexual\\nstage another (and placed under another class).  Later when the linkages\\nwere established, the fungi were correctly identified and moved out of\\ndeuteromycetes. Once perfect (sexual) stages of members of\\ndueteromycetes were discovered, they were often moved to ascomycetes\\nand basidiomycetes. The deuteromycetes reproduce only by asexual spores\\nknown as conidia. The mycelium is septate and branched. Some members\\nare saprophytes or parasites while a large number of them are\\ndecomposers of litter and help in mineral cycling. Some examples are\\nAlternaria, Colletotrichum and Trichoderma.\\nReprint 2025-26'),\n",
       " Document(metadata={'producer': 'GPL Ghostscript 8.15', 'creator': 'PageMaker 7.0', 'creationdate': '2022-09-19T14:27:46+00:00', 'author': 'Administrator', 'moddate': '2025-04-30T15:24:58+05:30', 'title': 'Organisms and Populations.pmd', 'source': './lebo111.pdf', 'total_pages': 17, 'page': 3, 'page_label': '4'}, page_content='191\\nORGANISMS AND POPULATIONS\\nYou have already learnt in previous classes that Ecology is a subject\\nwhich studies the interactions \\namong organisms and between the\\norganism and its physical (abiotic) environment.\\nEcology is basically concerned with four levels of biological\\norganisation  organisms, populations, communities and biomes.  In this\\nchapter we explore ecology at population levels.\\n11.1 POPULATIONS\\n11.1.1 Population Attributes\\nIn nature, we rarely find isolated, single individuals of any species; majority\\nof them live in groups in a well defined geographical area, share or compete\\nfor similar resources, potentially interbreed and thus constitute a\\npopulation. Although the term interbreeding implies sexual reproduction,\\na group of individuals resulting from even asexual reproduction is also\\ngenerally considered a population for the purpose of ecological studies.\\nAll the cormorants in a wetland, rats in an abandoned dwelling, teakwood\\ntrees in a forest tract, bacteria in a culture plate and lotus plants in a\\npond, are some examples of a population.  In earlier chapters you have\\nlearnt that although an individual organism is the one that has to cope\\nwith a changed environment, it is at the population level that natural\\nselection operates to evolve the desired traits. Population ecology is,\\ntherefore, an important area  because it links ecology to population genetics\\nand evolution.\\nA population has certain attributes whereas, an individual organism\\ndoes not. An individual may have births and deaths, but a population has\\nbirth rates and death rates. In a population these rates refer to per capita\\nbirths and deaths. The rates, hence, expressed are change in numbers\\n(increase or decrease) with respect to members of the population. Here is an\\nexample. If in a pond there were 20 lotus plants last year and through\\nreproduction 8 new plants are added, taking the current population to 28,\\nwe calculate the birth rate as 8/20 = 0.4 offspring per lotus per year. If 4\\nindividuals in a laboratory population of 40 fruitflies died during a specified\\ntime interval, say a week, the death rate in the population during that period\\nis 4/40 = 0.1 individuals per fruitfly per week.\\nAnother attribute characteristic of a population is sex ratio . An\\nindividual is either a male or a female but a population has a sex ratio\\n(e.g., 60 per cent of the population are females and 40 per cent males).\\nA population at any given time is composed of individuals of\\ndifferent ages. If the age distribution (per cent individuals of a given\\nage or age group) is plotted for the population, the resulting structure\\nis called an age pyramid (Figure 11.1). For human population, the\\nage pyramids generally show age distribution of males and females in\\na  diagram. The shape of the pyramids reflects the growth status of\\nthe population - (a) whether it is growing, (b) stable or (c) declining .\\nReprint 2025-26'),\n",
       " Document(metadata={'producer': 'GPL Ghostscript 8.15', 'creator': 'PageMaker 7.0', 'creationdate': '2022-10-14T14:18:24+00:00', 'author': 'Administrator', 'moddate': '2025-04-30T15:23:29+05:30', 'title': 'ch-3.pmd', 'source': './kebo103.pdf', 'total_pages': 14, 'page': 1, 'page_label': '2'}, page_content='24 BIOLOGY\\nnot only the external features, but also internal features, like ultra-\\nstructure, anatomy, embryology and phytochemistry. Such a\\nclassification for flowering plants was given by George Bentham and\\nJoseph Dalton Hooker.\\nAt present  phylogenetic classification systems based on\\nevolutionary relationships between the various organisms are acceptable.\\nThis assumes that organisms belonging to the same taxa have a common\\nancestor.   We now use information from many other sources too to help\\nresolve difficulties in classification.  These become more important when\\nthere is no supporting fossil evidence. Numerical Taxonomy which is\\nnow easily carried out using computers is  based on all observable\\ncharacteristics. Number and codes are assigned to all the characters and\\nthe data are then processed. In this way each character is given equal\\nimportance and at the same time hundreds of characters can be\\nconsidered. Cytotaxonomy, that is based on cytological information\\nlike chromosome number, structure, behaviour and chemotaxonomy,\\nthat uses the chemical constituents of the plant to resolve confusions,\\nare also used by taxonomists these days.\\n3.1 ALGAE\\nAlgae are chlorophyll-bearing, simple, thalloid, autotrophic and\\nlargely aquatic (both fresh water and marine) organisms. They occur\\nin a variety of other habitats: moist stones, soils and wood. Some of\\nthem also occur in association with fungi (lichen) and animals (e.g.,\\non sloth bear).\\nThe form and size of algae is highly variable, ranging from colonial\\nforms like Volvox and the filamentous forms like Ulothrix and Spirogyra\\n(Figure 3.1). A few of the marine forms such as kelps, form massive plant\\nbodies.\\nThe algae reproduce by vegetative, asexual and sexual methods.\\nVegetative reproduction is by fragmentation. Each fragment develops into\\na thallus. Asexual reproduction is by the production of different types of\\nspores, the most common being the zoospores. They are flagellated\\n(motile) and on germination gives rise to new plants. Sexual reproduction\\ntakes place through fusion of two gametes. These gametes can be\\nflagellated and similar in size (as in Ulothrix) or non-flagellated (non-motile)\\nbut similar in size (as in Spirogyra). Such reproduction is called\\nisogamous. Fusion of two gametes dissimilar in size, as in species of\\nEudorina is termed as anisogamous. Fusion between one large, non-\\nmotile (static) female gamete and a smaller, motile male gamete is termed\\noogamous, e.g., Volvox, Fucus.\\nReprint 2025-26'),\n",
       " Document(metadata={'producer': 'GPL Ghostscript 8.15', 'creator': 'PageMaker 7.0', 'creationdate': '2022-10-14T14:16:33+00:00', 'author': 'Administrator', 'moddate': '2025-04-30T15:23:27+05:30', 'title': 'ch-2.pmd', 'source': './kebo102.pdf', 'total_pages': 13, 'page': 9, 'page_label': '10'}, page_content='BIOLOGICAL CLASSIFICATION 19\\n2.4 KINGDOM PLANTAE\\nKingdom Plantae includes all eukaryotic chlorophyll-containing\\norganisms commonly called plants. A few members are partially\\nheterotrophic such as the insectivorous plants or parasites. Bladderwort\\nand Venus fly trap are examples of insectivorous plants and Cuscuta is a\\nparasite. The plant cells have an eukaryotic structure with prominent\\nchloroplasts and cell wall mainly made of cellulose. You will study the\\neukaryotic cell structure in detail in Chapter 8. Plantae includes algae,\\nbryophytes, pteridophytes, gymnosperms and angiosperms.\\nLife cycle of plants has two distinct phases  the diploid sporophytic\\nand the haploid gametophytic  that alternate with each other. The lengths\\nof the haploid and diploid phases, and whether these phases are free\\nliving or dependent on others, vary among different groups in plants.\\nThis phenomenon is called alternation of generations. You will study\\nfurther details of this kingdom in Chapter3.\\n2.5 KINGDOM ANIMALIA\\nThis kingdom is characterised by heterotrophic eukaryotic organisms\\nthat are multicellular and their cells lack cell walls. They directly or\\nindirectly depend on plants for food. They digest their food in an internal\\ncavity and store food reserves as glycogen or fat. Their mode of nutrition\\nis holozoic  by ingestion of food. They follow a definite growth pattern\\nand grow into adults that have a definite shape and size. Higher forms\\nshow elaborate sensory and neuromotor mechanism. Most of them are\\ncapable of locomotion.\\nThe sexual reproduction is by copulation of male and female followed\\nby embryological development. Salient features of various phyla are\\ndescribed in Chapter 4.\\n2.6 VIRUSES, VIROIDS, PRIONS AND LICHENS\\nIn the five kingdom classification of Whittaker there is no mention of lichens\\nand some acellular organisms like viruses, viroids and prions. These are\\nbriefly introduced here.\\nAll of us who have suffered the ill effects of common cold or flu know\\nwhat effects viruses can have on us, even if we do not associate it with our\\ncondition. Viruses did not find a place in classification since they are not\\nconsidered truly living, if we understand living as those organisms that\\nhave a cell structure.  The viruses are non-cellular organisms that are\\ncharacterised by having an inert crystalline structure outside the living cell.\\nReprint 2025-26')]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain.retrievers import BM25Retriever\n",
    "from langchain.retrievers import EnsembleRetriever\n",
    "import pickle\n",
    "\n",
    "with open(\"data.pkl\", \"rb\") as f:   # open file in binary read mode\n",
    "    docs = pickle.load(f, encoding=\"utf-8\")\n",
    "\n",
    "bm25_retriever = BM25Retriever.from_documents(docs)  # keyword-based\n",
    "bm25_retriever.k = 5\n",
    "\n",
    "vector_retriever = vector_store.as_retriever(search_kwargs={\"k\": 5})\n",
    "\n",
    "retriever = EnsembleRetriever(\n",
    "    retrievers=[bm25_retriever, vector_retriever],\n",
    "    weights=[0.5, 0.5],\n",
    ")\n",
    "\n",
    "retriever.invoke('examples of isogamous reproduction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "489fe280",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"OPENAI_API_KEY\"] = \"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\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75f78445",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm = ChatOpenAI(model=\"gpt-4o-mini\", temperature=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ed014e6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.prompts import PromptTemplate\n",
    "from langchain_core.runnables import RunnableParallel, RunnablePassthrough, RunnableLambda\n",
    "from langchain_core.output_parsers import StrOutputParser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8b32f021",
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt = PromptTemplate(\n",
    "    template=\"\"\"\n",
    "      You are a helpful teacher.\n",
    "      Answer ONLY from the provided transcript context.\n",
    "      If the context is insufficient, just say that this question is out of syllabus.\n",
    "      after your answer, also give the source and page label of the context, which is used.\n",
    "\n",
    "      {context}\n",
    "      Question: {question}\n",
    "    \"\"\",\n",
    "    input_variables = ['context', 'question']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3250ae98",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = StrOutputParser()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2cab0943",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_docs(retrieved_docs):\n",
    "  context_text = \"\\n\\n\".join(doc.page_content \n",
    "                             + ' \\nsource : ' \n",
    "                             + doc.metadata['source']\n",
    "                             + ' \\npage number : '\n",
    "                             + str(doc.metadata['page'])\n",
    "                             for doc in retrieved_docs)\n",
    "  return context_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2711cb38",
   "metadata": {},
   "outputs": [],
   "source": [
    "parallel_chain = RunnableParallel({\n",
    "    'context' : retriever | RunnableLambda(format_docs),\n",
    "    'question' : RunnablePassthrough(), \n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "cd7efb04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'context': 'HSEX\\nU\\nU\\nM\\nA\\nA\\nL\\nN\\nR REP\\nE\\nR\\nP\\nO\\nR\\nD\\nO\\nU\\nD\\nC\\nU\\nT\\nC\\nIO\\nT\\nN\\nIO\\nI\\nN\\nN FLOWERING PLANTS\\nvarieties some of the nucellar cells surrounding the embryo sac start\\ndividing, protrude into the embryo sac and develop into the embryos. In\\nsuch species each ovule contains many embryos. Occurrence of more\\nthan one embryo in a seed is referred to as polyembryony. Take out\\nsome seeds of orange and squeeze them. Observe the many embryos of\\ndifferent sizes and shapes from each seed. Count the number of embryos\\nin each seed. What would be the genetic nature of apomictic embryos?\\nCan they be called clones?\\nHybrid varieties of several of our food and vegetable crops are being\\nextensively cultivated. Cultivation of hybrids has tremendously increased\\nproductivity. One of the problems of hybrids is that hybrid seeds have\\nto be produced every year. If the seeds collected from hybrids are sown,\\nthe plants in the progeny will segregate and do not maintain hybrid\\ncharacters. Production of hybrid seeds is costly and hence the cost of\\nhybrid seeds become too expensive for the farmers. If these hybrids are\\nmade into apomicts, there is no segregation of characters in the hybrid\\nprogeny. Then the farmers can keep on using the hybrid seeds to raise\\nnew crop year after year and he does not have to buy hybrid seeds every\\nyear. Because of the importance of apomixis in hybrid seed industry,\\nactive research is going on in many laboratories around the world to\\nunderstand the genetics of apomixis and to transfer apomictic genes\\ninto hybrid varieties.\\nSUMMARY\\nFlowers are the seat of sexual reproduction in angiosperms. In the flower,\\nandroecium consisting of stamens represents the male reproductive\\norgans and gynoecium consisting of pistils represents the female\\nreproductive organs.\\nA typical anther is bilobed, dithecous and tetrasporangiate. Pollen\\ngrains develop inside the microsporangia. Four wall layers, the\\nepidermis, endothecium, middle layers and the tapetum surround the\\nmicrosporangium. Cells of the sporogenous tissue lying in the centre of\\nthe microsporangium, undergo meiosis (microsporogenesis) to form\\ntetrads of microspores. Individual microspores mature into pollen grains.\\nPollen grains represents the male gametophytic generation. The\\npollen grains have a two-layered wall, the outer exine and inner intine.\\nThe exine is made up of sporopollenin and has germ pores. Pollen grains\\nmay have two cells (a vegetative cell and generative cell) or three cells (a\\nvegetative cell and two male gametes) at the time of shedding.\\n23\\nThe pistil has three parts  the stigma, style and the ovary. Ovules\\nare present in the ovary. The ovules have a stalk called funicle, protective\\nintegument(s), and an opening called micropyle. The central tissue is\\nthe nucellus in which the archesporium differentiates. A cell of the\\narchesporium, the megaspore mother cell divides meiotically and one of\\nthe megaspores forms the embryo sac (the female gametophyte). The\\nmature embryo sac is 7-celled and 8-nucleate. At the micropylar end is\\nReprint 2025-26\\n \\nsource : ./lebo101.pdf \\npage number : 22\\n\\nBIOLOGY\\nlie close to each other so that self-pollination\\ncan occur. Some plants such as Viola\\n(common pansy), Oxalis, and Commelina\\nproduce two types of flowers \\nchasmogamous flowers which are similar to\\nflowers of other species with exposed anthers\\nand stigma, and cleistogamous flowers which\\ndo not open at all (Figure 1.9c). In such flowers,\\nthe anthers and stigma lie close to each other.\\n(a) When anthers dehisce in the flower buds,\\npollen grains come in contact with the stigma\\nto effect pollination. Thus, cleistogamous\\nflowers are invariably autogamous as there is\\nno chance of cross-pollen landing on the\\nstigma. Cleistogamous flowers produce\\nassured seed-set even in the absence of\\npollinators. Do you think that cleistogamy is\\nadvantageous or disadvantageous to the\\nplant? Why?\\n(b) (ii) GGGGGeeeeeiiiiitttttooooonnnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nthe anther to the stigma of another flower of\\nthe same plant. Although geitonogamy is\\nfunctionally cross-pollination involving a\\npollinating agent, genetically it is similar to\\nautogamy since the pollen grains come from\\nthe same plant.\\n(iii) XXXXXeeeeennnnnooooogggggaaaaammmmmyyyyy  Transfer of pollen grains from\\nanther to the stigma of a different plant (Figure\\n1.9b). This is the only type of pollination which\\nduring pollination brings genetically different\\ntypes of pollen grains to the stigma.\\nAgents of Pollination : Plants use two abiotic (wind\\nand water) and one biotic (animals) agents to achieve\\npollination. Majority of plants use biotic agents for\\npollination. Only a small proportion of plants use\\n12\\n(c) abiotic agents. Pollen grains coming in contact with\\nthe stigma is a chance factor in both wind and water\\npollination. To compensate for this uncertainties and\\nassociated loss of pollen grains, the flowers produce\\nFigure 1.9 (a) Self-pollinated flowers;\\n(b)Cross pollinated flowers; enormous amount of pollen when compared to the\\n(c)Cleistogamous flowers\\nnumber of ovules available for pollination.\\nReprint 2025-26\\n \\nsource : ./lebo101.pdf \\npage number : 11\\n\\nBIOLOGICAL CLASSIFICATION 17\\ncalled conidia or sporangiospores or zoospores, and sexual reproduction\\nis by oospores, ascospores and basidiospores. The various spores are\\nproduced in distinct structures called fruiting bodies. The sexual cycle\\ninvolves the following  three steps:\\n(i) Fusion of protoplasms between two motile or non-motile gametes\\ncalled plasmogamy.\\n(ii) Fusion of two nuclei called karyogamy.\\n(iii) Meiosis in zygote resulting in haploid spores.\\nWhen a fungus reproduces sexually, two haploid\\nhyphae of compatible mating types come together and\\nfuse. In some fungi the fusion of two haploid cells\\nimmediately results in diploid cells (2n). However, in other\\nfungi (ascomycetes and basidiomycetes), an intervening\\ndikaryotic stage (n + n, i.e., two nuclei per cell) occurs;\\nsuch a condition is called a dikaryon and the phase is\\ncalled dikaryophase of fungus. Later, the parental nuclei\\nfuse and the cells become diploid.  The fungi form fruiting\\nbodies in which reduction division occurs, leading to\\nformation of haploid spores.\\nThe morphology of the mycelium, mode of spore\\nformation and fruiting bodies  form the basis for the\\ndivision of the kingdom  into various classes.\\n2.3.1 Phycomycetes\\nMembers of phycomycetes are found in aquatic habitats\\nand on decaying wood in moist and damp places or as\\nobligate parasites on plants. The mycelium is aseptate\\nand coenocytic. Asexual reproduction takes place by\\nzoospores (motile) or by aplanospores (non-motile). These\\nspores are endogenously produced in sporangium. A\\nzygospore is formed by fusion of two gametes. These\\ngametes are similar in morphology (isogamous) or\\ndissimilar (anisogamous or oogamous). Some common\\nexamples are Mucor (Figure 2.5a), Rhizopus (the bread\\nmould mentioned earlier) and Albugo (the parasitic fungi\\non mustard).\\n2.3.2 Ascomycetes\\nCommonly known as sac-fungi, the ascomycetes  are mostly\\nmulticellular, e.g., Penicillium, or rarely unicellular, e.g., yeast\\n(Saccharomyces). They are saprophytic, decomposers,\\nparasitic or coprophilous (growing on dung). Mycelium\\nFigure 2.5 Fungi: (a) Mucor\\n(b) Aspergillus (c) Agaricus\\n(c)\\n(a)\\n(b)\\nReprint 2025-26 \\nsource : ./kebo102.pdf \\npage number : 7\\n\\n28 BIOLOGY\\ngametes. Sexual reproduction is oogamous and accompanied by complex\\npost fertilisation developments. The common members are: Polysiphonia,\\nPorphyra (Figure 3.1c), Gracilaria and Gelidium.\\n3.2 BRYOPHYTES\\nBryophytes include the various mosses and liverworts that are found\\ncommonly growing in moist shaded areas in the hills (Figure 3.2).\\nArchegoniophore\\n(a) (b)\\n(c)\\n(d)\\nAntheridiophore\\nCapsule\\nAntheridial\\nbranch Branches\\nArchegonial\\nbranch\\nSeta\\nSporophyte\\nGametophyte\\nLeaves\\nMain axis\\nRhizoids\\nGemma cup\\nRhizoids\\nGemma cup\\nRhizoids\\nFigure 3.2 Bryophytes: A liverwort  Marchantia (a) Female thallus  (b) Male thallus\\nMosses   (c) Funaria, gametophyte and sporophyte  (d) Sphagnum\\ngametophyte\\nReprint 2025-26 \\nsource : ./kebo103.pdf \\npage number : 5\\n\\n18 BIOLOGY\\nis branched and septate. The asexual spores are conidia produced\\nexogenously on the special mycelium called conidiophores. Conidia on\\ngermination produce mycelium. Sexual spores are called ascospores\\nwhich are produced endogenously in sac like asci (singular ascus). These\\nasci are arranged in different types of fruiting bodies called ascocarps.\\nSome examples are Aspergillus (Figure 2.5b), Claviceps and Neurospora.\\nNeurospora  is used extensively in biochemical and genetic work. Many\\nmembers like morels and truffles are edible and are considered  delicacies.\\n2.3.3 Basidiomycetes\\nCommonly known forms of basidiomycetes are mushrooms, bracket\\nfungi or puffballs. They grow in soil, on logs and tre e stumps and in\\nliving plant bodies as parasites, e.g., rusts and smuts. The mycelium is\\nbranched and septate. The asexual spores are generally not found, but\\nvegetative reproduction by fragmentation is common. The sex organs\\nare absent, but plasmogamy is brought about by fusion of two vegetative\\nor somatic cells of different strains or genotypes. The resultant structure\\nis dikaryotic which ultimately gives rise to basidium. Karyogamy and\\nmeiosis take place in the basidium producing four basidiospores. The\\nbasidiospores are exogenously produced on the basidium (pl.: basidia).\\nThe basidia are arranged in fruiting bodies called basidiocarps. Some\\ncommon members are Agaricus (mushroom) (Figure 2.5c), Ustilago \\n(smut)\\nand Puccinia (rust fungus).\\n2.3.4 Deuteromycetes\\nCommonly known as imperfect fungi because only the asexual or\\nvegetative phases of these fungi are known.  When the sexual forms of\\nthese fungi were discovered they were moved into classes they rightly\\nbelong to.  It is also possible that the asexual and vegetative stage have\\nbeen given one name (and placed under deuteromycetes) and the sexual\\nstage another (and placed under another class).  Later when the linkages\\nwere established, the fungi were correctly identified and moved out of\\ndeuteromycetes. Once perfect (sexual) stages of members of\\ndueteromycetes were discovered, they were often moved to ascomycetes\\nand basidiomycetes. The deuteromycetes reproduce only by asexual spores\\nknown as conidia. The mycelium is septate and branched. Some members\\nare saprophytes or parasites while a large number of them are\\ndecomposers of litter and help in mineral cycling. Some examples are\\nAlternaria, Colletotrichum and Trichoderma.\\nReprint 2025-26 \\nsource : ./kebo102.pdf \\npage number : 8\\n\\n191\\nORGANISMS AND POPULATIONS\\nYou have already learnt in previous classes that Ecology is a subject\\nwhich studies the interactions \\namong organisms and between the\\norganism and its physical (abiotic) environment.\\nEcology is basically concerned with four levels of biological\\norganisation  organisms, populations, communities and biomes.  In this\\nchapter we explore ecology at population levels.\\n11.1 POPULATIONS\\n11.1.1 Population Attributes\\nIn nature, we rarely find isolated, single individuals of any species; majority\\nof them live in groups in a well defined geographical area, share or compete\\nfor similar resources, potentially interbreed and thus constitute a\\npopulation. Although the term interbreeding implies sexual reproduction,\\na group of individuals resulting from even asexual reproduction is also\\ngenerally considered a population for the purpose of ecological studies.\\nAll the cormorants in a wetland, rats in an abandoned dwelling, teakwood\\ntrees in a forest tract, bacteria in a culture plate and lotus plants in a\\npond, are some examples of a population.  In earlier chapters you have\\nlearnt that although an individual organism is the one that has to cope\\nwith a changed environment, it is at the population level that natural\\nselection operates to evolve the desired traits. Population ecology is,\\ntherefore, an important area  because it links ecology to population genetics\\nand evolution.\\nA population has certain attributes whereas, an individual organism\\ndoes not. An individual may have births and deaths, but a population has\\nbirth rates and death rates. In a population these rates refer to per capita\\nbirths and deaths. The rates, hence, expressed are change in numbers\\n(increase or decrease) with respect to members of the population. Here is an\\nexample. If in a pond there were 20 lotus plants last year and through\\nreproduction 8 new plants are added, taking the current population to 28,\\nwe calculate the birth rate as 8/20 = 0.4 offspring per lotus per year. If 4\\nindividuals in a laboratory population of 40 fruitflies died during a specified\\ntime interval, say a week, the death rate in the population during that period\\nis 4/40 = 0.1 individuals per fruitfly per week.\\nAnother attribute characteristic of a population is sex ratio . An\\nindividual is either a male or a female but a population has a sex ratio\\n(e.g., 60 per cent of the population are females and 40 per cent males).\\nA population at any given time is composed of individuals of\\ndifferent ages. If the age distribution (per cent individuals of a given\\nage or age group) is plotted for the population, the resulting structure\\nis called an age pyramid (Figure 11.1). For human population, the\\nage pyramids generally show age distribution of males and females in\\na  diagram. The shape of the pyramids reflects the growth status of\\nthe population - (a) whether it is growing, (b) stable or (c) declining .\\nReprint 2025-26 \\nsource : ./lebo111.pdf \\npage number : 3\\n\\n24 BIOLOGY\\nnot only the external features, but also internal features, like ultra-\\nstructure, anatomy, embryology and phytochemistry. Such a\\nclassification for flowering plants was given by George Bentham and\\nJoseph Dalton Hooker.\\nAt present  phylogenetic classification systems based on\\nevolutionary relationships between the various organisms are acceptable.\\nThis assumes that organisms belonging to the same taxa have a common\\nancestor.   We now use information from many other sources too to help\\nresolve difficulties in classification.  These become more important when\\nthere is no supporting fossil evidence. Numerical Taxonomy which is\\nnow easily carried out using computers is  based on all observable\\ncharacteristics. Number and codes are assigned to all the characters and\\nthe data are then processed. In this way each character is given equal\\nimportance and at the same time hundreds of characters can be\\nconsidered. Cytotaxonomy, that is based on cytological information\\nlike chromosome number, structure, behaviour and chemotaxonomy,\\nthat uses the chemical constituents of the plant to resolve confusions,\\nare also used by taxonomists these days.\\n3.1 ALGAE\\nAlgae are chlorophyll-bearing, simple, thalloid, autotrophic and\\nlargely aquatic (both fresh water and marine) organisms. They occur\\nin a variety of other habitats: moist stones, soils and wood. Some of\\nthem also occur in association with fungi (lichen) and animals (e.g.,\\non sloth bear).\\nThe form and size of algae is highly variable, ranging from colonial\\nforms like Volvox and the filamentous forms like Ulothrix and Spirogyra\\n(Figure 3.1). A few of the marine forms such as kelps, form massive plant\\nbodies.\\nThe algae reproduce by vegetative, asexual and sexual methods.\\nVegetative reproduction is by fragmentation. Each fragment develops into\\na thallus. Asexual reproduction is by the production of different types of\\nspores, the most common being the zoospores. They are flagellated\\n(motile) and on germination gives rise to new plants. Sexual reproduction\\ntakes place through fusion of two gametes. These gametes can be\\nflagellated and similar in size (as in Ulothrix) or non-flagellated (non-motile)\\nbut similar in size (as in Spirogyra). Such reproduction is called\\nisogamous. Fusion of two gametes dissimilar in size, as in species of\\nEudorina is termed as anisogamous. Fusion between one large, non-\\nmotile (static) female gamete and a smaller, motile male gamete is termed\\noogamous, e.g., Volvox, Fucus.\\nReprint 2025-26 \\nsource : ./kebo103.pdf \\npage number : 1\\n\\nBIOLOGICAL CLASSIFICATION 19\\n2.4 KINGDOM PLANTAE\\nKingdom Plantae includes all eukaryotic chlorophyll-containing\\norganisms commonly called plants. A few members are partially\\nheterotrophic such as the insectivorous plants or parasites. Bladderwort\\nand Venus fly trap are examples of insectivorous plants and Cuscuta is a\\nparasite. The plant cells have an eukaryotic structure with prominent\\nchloroplasts and cell wall mainly made of cellulose. You will study the\\neukaryotic cell structure in detail in Chapter 8. Plantae includes algae,\\nbryophytes, pteridophytes, gymnosperms and angiosperms.\\nLife cycle of plants has two distinct phases  the diploid sporophytic\\nand the haploid gametophytic  that alternate with each other. The lengths\\nof the haploid and diploid phases, and whether these phases are free\\nliving or dependent on others, vary among different groups in plants.\\nThis phenomenon is called alternation of generations. You will study\\nfurther details of this kingdom in Chapter3.\\n2.5 KINGDOM ANIMALIA\\nThis kingdom is characterised by heterotrophic eukaryotic organisms\\nthat are multicellular and their cells lack cell walls. They directly or\\nindirectly depend on plants for food. They digest their food in an internal\\ncavity and store food reserves as glycogen or fat. Their mode of nutrition\\nis holozoic  by ingestion of food. They follow a definite growth pattern\\nand grow into adults that have a definite shape and size. Higher forms\\nshow elaborate sensory and neuromotor mechanism. Most of them are\\ncapable of locomotion.\\nThe sexual reproduction is by copulation of male and female followed\\nby embryological development. Salient features of various phyla are\\ndescribed in Chapter 4.\\n2.6 VIRUSES, VIROIDS, PRIONS AND LICHENS\\nIn the five kingdom classification of Whittaker there is no mention of lichens\\nand some acellular organisms like viruses, viroids and prions. These are\\nbriefly introduced here.\\nAll of us who have suffered the ill effects of common cold or flu know\\nwhat effects viruses can have on us, even if we do not associate it with our\\ncondition. Viruses did not find a place in classification since they are not\\nconsidered truly living, if we understand living as those organisms that\\nhave a cell structure.  The viruses are non-cellular organisms that are\\ncharacterised by having an inert crystalline structure outside the living cell.\\nReprint 2025-26 \\nsource : ./kebo102.pdf \\npage number : 9',\n",
       " 'question': 'examples of isogamous reproduction'}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parallel_chain.invoke('examples of isogamous reproduction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "afb7e749",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_chain = parallel_chain | prompt | llm | parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e897c748",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Examples of isogamous reproduction include the fusion of two gametes that are similar in size, as seen in Ulothrix. \\n\\nSource: ./kebo103.pdf, page number: 1'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "answer = main_chain.invoke('examples of isogamous reproduction')\n",
    "answer"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lang",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
